---
title: "Hannum 2013 Epigenetic Clock Implementation"
author: "Fabien Jossaud, Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---


```{r echo=FALSE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
options(knitr.duplicate.label = "allow")
```

# Purpose

DNA methylation consists in adding a methyl group to the DNA molecule. 
It is an egigentic mechanism that is used by cells to control gene expression.
The most abundant and studied DNA methylation is the methylation of a Cythosine followed by a Guanine (CpG dicucleotide).

Since the maintenance of an epigenetic state is an active process mediated by the metabolic activity of cells and this metabolic activity evolves over the life of an individual: 
i) DNA methylation becomes a good predictor of aging
ii) aging becomes a good explanatory factor of DNA methylation, allowing to capture the labile part of the epigenome.


Changes in DNA methylation during aging have been widely studied [Yousefi NRG 2022] and formalized by *epigenetic clock theory* [Hannum Mol. Cell 2013, Horvath NRG 2018], revealing *e.g.* placenta specific gene reactivation [Liu Dev. Cell 2022] as it has ever been pointed out in cancer [Rousseaux Sci. Tr. Med. 2013].

Our hypothesis is that the epigenetic hotspots highlighted by the study of aging are good candidate biomarkers of the risk associated with exposure to environmental polluants or of developing a chronic disease.

Studying meth with regard to age makes it possible to capture the plastic part of mtDNA


**Biological clocks** are statstical tools allowing to **predict** age according bilogical parameters.

*e.g.*

$$Age \sim telomere size$$ [Ref.]

$$Age \sim composition cellulaire$$ 

$$Age \sim DNAm$$


Note: Predictive model are different than explanatory model.

*e.g.*

15% of 450k probes are correlated to the age,
but only 71 probes are used in Hannum 2013 clock.


Since, we use the clock metaphore, aging characterizes a shift between chronological age and biologiocal (or predicted) age.
But, more than aging, epigenetic clock points out the DNAm plastic part over time.

Then, we use epigenetic clock (predictiv model) to study cofactor effects (altitude,  air pollution...) on DNAm.
Thereby, epigenetic clock becomes a powerfull ligthweight tool to emphasis effects of cofactor on (15% of) DNAm.

*e.g.* 

In Hannum 2013, Aging Methylation Acceleration Rate (AMAR) according do sex shows differential aging.




Lightness of epigenetic clocks allow to treat cases more powerfully than it could be with DMR.

*e.g.*

clock on 27k  vs. DMR on 27k


















# Descriptive Statistics

```{r child = "02_statdesc.Rmd", eval=TRUE}
```

# Data Preprocessing

```{r child = "03_preproc.Rmd", eval=TRUE}
```

# Build Model


```{r child = "04_model.Rmd", eval=TRUE}
```


# Session Information

```{r, results="verbatim"}
sessionInfo()
```

