---
title: "Build Illumina methylation study from IDAT from NCBI GEO using RnBeads"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---


```{r echo=FALSE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
info = list(start_time = Sys.time())
```

```{r}
working_dir = getwd()
datashare_dir = "datashare"
idat_dir = paste0(datashare_dir, "/", gse, "/raw")
```


# Get IDAT from GEO

```{r download_gse_raw_tar}
if (substr(gse, 1, 3)=="GSE") {epimedtools::download_gse_raw_tar(gse, datashare_dir=datashare_dir)}
```

# Process IDAT

```{r prerocess_idat}
# unzip idat.gz files
setwd(idat_dir)
cmd = "gzip"
args = ("-d *.idat.gz")
print(paste(cmd, args))
system2(cmd, args)
setwd(working_dir)

# filtering idat files by sentrix of interest (if multiple infinium platforms)
detect_infinium_platform = function (idat_filename) { # inspired by RnBeads:::rnb.detect.infinium.platform
  print(idat_filename)
  inf27k.idats.present = any(grepl("_[ABCDEFGHIJKL]_", 
      idat_filename))
  inf450kEPIC.idats.present = any(grepl("_R0[1-9]C0[1-2]_", 
      idat_filename))
  if (inf27k.idats.present) {
      if (inf450kEPIC.idats.present) {
          stop("Undefined platform; detected HumanMethylation27 and HumanMethylation450")
      }
      return("GPL8490")
  }
  if (inf450kEPIC.idats.present) {
      file.sizes = as.numeric(na.omit(file.info(idat_filename)[, 
          "size"]))
      if (length(file.sizes) == 0) {
          stop("Undefined platform; cannot read the specified IDAT files")
      }
      if (all(file.sizes > 1e+07)) {
          return("GPL21145")
      }
      if (all(file.sizes < 1e+07)) {
          return("GPL13534")
      }
      stop("Undefined platform; detected HumanMethylation450 and MethylationEPIC")
  }
  stop("Undefined platform; unexpected or missing IDAT files")
}
idat_red_files = list.files(idat_dir, "*_Red.idat")
if (substr(gse, 1, 3)=="GSE") {idat_red_files = idat_red_files[substr(idat_red_files, 1, 3)=="GSM"]}
# idat_red_files = do.call(rbind, strsplit(idat_red_files, "_"))[,-1]
# idat_red_files = apply(idat_red_files, 1, paste, collapse="_")
setwd(idat_dir)
infinium_platform = sapply(idat_red_files, detect_infinium_platform)
tmp_tab = table(infinium_platform)
if (length(tmp_tab) > 1) {
  pf_idx = names(rev(sort(tmp_tab)))[1]
  idat_red_files = idat_red_files[infinium_platform==pf_idx]
  infinium_platform = sapply(idat_red_files, detect_infinium_platform)
  tmp_tab = table(infinium_platform)
}
if (substr(gse, 1, 3)=="GSE") {
  sentrix_oi = unique(do.call(rbind, strsplit(idat_red_files, "_"))[,2])
} else {
  sentrix_oi = unique(do.call(rbind, strsplit(idat_red_files, "_"))[,1])  
}
pf_idx = names(tmp_tab)
setwd(working_dir)


# Design
idat_red_files = list.files(idat_dir, "*_Red.idat")
if (substr(gse, 1, 3)=="GSE") {idat_red_files = idat_red_files[substr(idat_red_files, 1, 3)=="GSM"]}
if (substr(gse, 1, 3)=="GSE") {
  exp_grp = do.call(rbind, strsplit(idat_red_files, "_"))[,1:3]
} else {
  exp_grp = do.call(rbind, strsplit(idat_red_files, "_"))[,1:2]
  exp_grp = cbind(paste0("X", exp_grp[,1], "_", exp_grp[,2]), exp_grp)  
}
colnames(exp_grp) = c("Sample_ID", "Sentrix_ID", "Sentrix_Position")
rownames(exp_grp) = exp_grp[,1]
exp_grp = as.data.frame(exp_grp)
dim(exp_grp)
exp_grp = exp_grp[exp_grp$Sentrix_ID%in%sentrix_oi,]
dim(exp_grp)
# exp_grp = exp_grp[1:500,]
head(exp_grp)
dim(exp_grp)
exp_grp_orig = exp_grp
```


```{r}
# format idat file using symblink
if (substr(gse, 1, 3)=="GSE") {
  setwd(idat_dir)
  foo = apply(exp_grp, 1, function (row) {
    cmd = "ln"
    args = paste0("-s ", row[[1]], "_", row[[2]], "_", row[[3]], "_Red.idat ", row[[2]], "_", row[[3]], "_Red.idat")
    print(paste(cmd, args))
    system2(cmd, args)
    args = paste0("-s ", row[[1]], "_", row[[2]], "_", row[[3]], "_Grn.idat ", row[[2]], "_", row[[3]], "_Grn.idat")
    print(paste(cmd, args))
    system2(cmd, args)
  })
  setwd(working_dir)
}
```


```{r rnb.options}
# Options, https://rdrr.io/bioc/RnBeads/man/rnb.options.html
library(RnBeads)
library(grid)
RnBeads::rnb.options(
  analysis.name        = gse,
  email = "florent.chuffart@univ-grenoble-alpes.fr",
  filtering.greedycut=FALSE,
  filtering.snp="no",
  filtering.cross.reactive=FALSE,
  filtering.sex.chromosomes.removal=FALSE, 
  identifiers.column="Sample_ID",
  export.to.csv=TRUE,
  # replicate.id.column = "treatment",
  # import.table.separator = ";",
  # import.bed.style     = "bismarkCov",
  # assembly             = version,
  # region.types=c("promoters", "genes", "tiling", "cpgislands", "sites"),
  region.types=c("sites"),
  # region.aggregation="coverage.weighted",
  gz.large.files=TRUE,
  # differential.enrichment.go = TRUE
  # differential.enrichment.lola = TRUE,

  # qc.barplots=FALSE,
  # qc.sample.batch.size = 50,

  qc.boxplots	= FALSE        , 
  qc.barplots	= FALSE        , 
  qc.negative.boxplot	= FALSE, 
  qc.snp.heatmap	= FALSE    , 
  qc.snp.barplot	= FALSE    , 
  qc.snp.boxplot	= FALSE    , 
  qc.snp.distances	= FALSE  , 
  qc.snp.purity	= FALSE      , 
  qc.cnv	= FALSE            , 
  qc.cnv.refbased	= FALSE    , 

  export.to.bed        = FALSE,
  export.to.trackhub   = NULL,
  # export.to.ewasher    = "no",

  exploratory=FALSE,
  differential=FALSE
)
# foo = RnBeads::rnb.options()
# foo$export.to.trackhub
```


```{r RnBeads}
exp_grp = exp_grp_orig
dim(exp_grp)
# exp_grp = exp_grp[c(1:10,1:10),]
# dim(exp_grp)

if (pf_idx=="GPL21145") {
  nb_sample_max = 2477 
} else if (pf_idx=="GPL13534") {
  nb_sample_max = 5000  
} else {
  nb_sample_max = 80000    
}

nb_batch = ceiling(nrow(exp_grp) / 2477)
nb_sentrix = length(unique(exp_grp$Sentrix_ID))
set.seed(1)
batch_sentrix_dict = sample(1:nb_batch, nb_sentrix, replace=TRUE)
names(batch_sentrix_dict) = unique(exp_grp$Sentrix_ID)
exp_grp$batch = as.factor(paste0("batch", batch_sentrix_dict[exp_grp$Sentrix_ID]))
head(exp_grp)
table(exp_grp$batch)


dfs = lapply(levels(exp_grp$batch), function(batch) {
  # exp_grp in ====>
  # design (sample.sheet)
  sample.sheet = paste0("exp_grp_", batch, ".csv")
  tmp_exp_grp = exp_grp[exp_grp$batch==batch,]
  print(dim(tmp_exp_grp))
  head(tmp_exp_grp)
  write.table(tmp_exp_grp, sample.sheet, sep=",", row.names=FALSE, quote=FALSE)
  # Directory where the output should be written to
  analysis.dir = paste0(working_dir, "/", gse, "/analysis_", batch)

  # GO!
  dir.reports = file.path(analysis.dir, "reports")
  data.type="infinium.idat.dir"
  dir.create(dir.reports, showWarnings=FALSE)
  num.cores = parallel::detectCores()
  parallel.setup(num.cores)
  unlink(dir.reports, recursive=TRUE)
  RnBeads::rnb.run.analysis(
    # GEO.acc = gse,
    sample.sheet=sample.sheet,
    data.dir=idat_dir,
    # data.source = list(data.dir = idat_dir, sample.sheet = sample.sheet),
    # GS.report = NULL,

    dir.reports=dir.reports, 
    # build.index = TRUE,
    save.rdata = FALSE,
    initialize.reports=TRUE
  )
 
  # read output
  beta_matrixcsv_file = paste0(analysis.dir, "/reports/tracks_and_tables_data/csv/betas_1.csv.gz")
  print(beta_matrixcsv_file)
  res = read.table(beta_matrixcsv_file, sep=",", row.names=1, header=TRUE)
  head(res[,1:10])
  dim(res)
  return(res)
})


idx_probes = rownames(dfs[[1]])
if (length(dfs) >= 2) {
  for (i in 2:length(dfs)) {
    idx_probes = intersect(idx_probes, rownames(dfs[[i]]))
  }  
}

sapply(dfs, dim)
length(idx_probes)

df = dfs[[1]][idx_probes, ]
if (length(dfs) >= 2) {
  for (i in 2:length(dfs)) {
    df = cbind(df, dfs[[i]][idx_probes, -(1:4)])
  }  
}

sapply(dfs, dim)
length(idx_probes)
dim(df)
head(df[,1:10])
rownames(exp_grp)[!rownames(exp_grp) %in% colnames(df)]

# ===> df out
head(df[,1:10])
dim(df)
```








# Save

```{r save,  echo=TRUE, results="verbatim"}
pf = df[,1:4]
head(pf)
dim(pf)

exp_grp = exp_grp[intersect(rownames(exp_grp), colnames(df)),]
head(exp_grp)
dim(exp_grp)

data = as.matrix(df[,rownames(exp_grp)])
head(data[,1:6])
dim(data)

s = epimedtools::create_study()
s$data     = data
s$exp_grp  = exp_grp
s$platform = pf
s$platform_name = pf_idx
s_filename = paste0("datashare/", gse, "/study_idat_", gse, ".rds")
print(paste0("Writing ", s_filename, "..."))
s$save(s_filename)
```

# Process Information 

```{r info, results="verbatim"}
info$stop_time = Sys.time()
info$n   = nrow(s$data)
info$p   = ncol(s$data)
info$stop_timeexec_time = difftime(info$stop_time, info$start_time, units="mins")
print(info)
saveRDS(info, paste0("info_idat2study_",gse,".rds"))
```


# Session Information

```{r, results="verbatim"}
sessionInfo()
```






```{r eval=FALSE}
# cd datashare/GSE197678/raw

dim(s$exp_grp)
s$exp_grp$key = substr(s$exp_grp$title, 77, 79)
tab = sort(table(s$exp_grp$key))
tab
thresh = 150
sample_idx = rownames(s$exp_grp)[s$exp_grp$key %in% names(tab)[tab >= thresh]]
# sample_idx = rownames(s$exp_grp)[s$exp_grp$key %in% names(tab)[tab < 10]]
length(sample_idx)

length(sample_idx) * 850000
.Machine$integer.max

for (gsm in sample_idx) {
  print(gsm)
  cmd = "mv"
  args = paste0("datashare/GSE197678/raw/back/", gsm, "*.idat datashare/GSE197678/raw/.")
  print(paste(cmd, args))
  system2(cmd, args)
}



for (i in 2480:2000) {
  print(paste0("*************** ", i, " ************"))
  exp_grp = exp_grp[1:i,]
  dim(exp_grp)
  write.table(exp_grp, sample.sheet, sep=",", row.names=FALSE, quote=FALSE)


  dir.create(dir.reports, showWarnings=FALSE)
  num.cores = 32
  parallel.setup(num.cores)

  unlink(dir.reports, recursive=TRUE)
  rnbtry = try(RnBeads::rnb.run.analysis(
    # GEO.acc = gse,
    sample.sheet=sample.sheet,
    data.dir=idat_dir,
    # data.source = list(data.dir = idat_dir, sample.sheet = sample.sheet),
    # GS.report = NULL,

    dir.reports=dir.reports, 
    # build.index = TRUE,
    save.rdata = FALSE,
    initialize.reports=TRUE
  ))
  if (!inherits(rnbtry, "try-error")) {
    stop() 
  } else {
    print(paste0("*************** NOT ", i, " ************"))
  }
}
```



