---
title: "ENCODE profiles"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---



```{r echo=FALSE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
info = list(start_time = Sys.time())
source("common.R")
# if (!exists("mreadRDS")) {mreadRDS = memoise::memoise(readRDS, cache=cachem::cache_mem(max_size = 10*1024 * 1024^2)) }
```

# ENCODE / deeptools


### Marks Associated with Transcriptional Activation:
1. **H3K27ac**  
   - Associated with active enhancers and promoters.  
   - Indicates a transcriptionally permissive environment.

2. **H3K36me3**  
   - Found in coding regions of actively transcribed genes.  
   - Linked to transcriptional elongation and genomic integrity.

3. **H3K4me1**  
   - Associated with enhancers (often primed or active).  
   - Reflects regulatory potential but does not guarantee activity.

4. **H3K4me3**  
   - Marks active or poised promoters.  
   - Strongly linked to transcriptional initiation.

5. **H3K9ac**  
   - Found at active promoters and enhancers.  
   - Associated with euchromatic, transcriptionally permissive environments.

---

### Marks Associated with Transcriptional Repression:
6. **H3K27me3**  
   - Linked to transcriptional repression, often mediated by the Polycomb Repressive Complex 2 (PRC2).  
   - Characteristic of facultative heterochromatin (regions that are repressed but potentially reactivatable).

7. **H3K9me3**  
   - A mark of constitutive heterochromatin, found in inactive genomic regions such as centromeres and telomeres.  
   - Recruits proteins like HP1 (Heterochromatin Protein 1) to maintain a condensed chromatin state.

---

### Summary:


| **Histone Mark** | **Activation/Repression**   | **Associated Regions**               |
|-------------------|-----------------------------|-------------------------------------|
| **H3K27ac**       | Activation                 | Active enhancers, promoters          |
| **H3K36me3**      | Activation                 | Coding regions                       |
| **H3K4me1**       | Activation (potential)     | Primed/active enhancers              |
| **H3K4me3**       | Activation                 | Active or poised promoters           |
| **H3K9ac**        | Activation                 | Active promoters, enhancers          |
| **H3K27me3**      | Repression                 | Facultative heterochromatin  (PRC2)  |
| **H3K9me3**       | Repression                 | Constitutive heterochromatin (HP1 )  |





```{r encode_db}
# db = read.table("~/projects/meth3d/results/entex/metadata.insituhic.tsv", sep="\t", header=TRUE)
db = read.table("encode_lung.tsv", sep="\t", header=TRUE)
# db = read.table("encode_ovary.tsv", sep="\t", header=TRUE)
rownames(db) = db[,1]
dim(db)
head(db)
table(db[,3])

table(db$File.format, db$File.type)
db = db[db$File.type == "bigWig",]

table(db$Experiment.target)
db$Experiment.target = factor(db$Experiment.target, levels=c(
  "H3K4me3-human"  , # Activation            | Active or poised promoters           |
  "H3K4me1-human"  , # Activation (potential)| Primed/active enhancers              |
  "H3K27me3-human" , # Repression            | Facultative heterochromatin  (PRC2)  |
  "H3K27ac-human"  , # Activation            | Active enhancers, promoters          |
  "H3K9ac-human"   , # Activation            | Active promoters, enhancers          |
  "H3K36me3-human" , # Activation            | Coding regions                       |
  "H3K9me3-human"  , # Repression            | Constitutive heterochromatin (HP1 )  |
  "CTCF-human", 
  NULL
))
db = db[order(db$Experiment.target),]
db = db[db$Experiment.target %in% c(
  "H3K27me3-human" , # Repression            | Facultative heterochromatin  (PRC2)  |
  "H3K4me3-human"  , # Activation            | Active or poised promoters           |
  "H3K4me1-human"  , # Activation (potential)| Primed/active enhancers              |
  # "H3K27ac-human"  , # Activation            | Active enhancers, promoters          |
  # "H3K9ac-human"   , # Activation            | Active promoters, enhancers          |
  # "H3K36me3-human" , # Activation            | Coding regions                       |
  # "H3K9me3-human"  , # Repression            | Constitutive heterochromatin (HP1 )  |
  NULL
),]

blacklist = c(
  # "ENCFF331IOZ", #     H3K27ac-human                lung
  # "ENCFF028TDS", #     H3K27ac-human                lung
  # "ENCFF386ZSF", #    H3K27me3-human                lung
  # "ENCFF238XGJ", #    H3K27me3-human                lung
  "ENCFF440HPT", #    H3K27me3-human                lung
  # "ENCFF507BZS", #    H3K36me3-human                lung
  # "ENCFF170SDA", #    H3K36me3-human                lung
  # "ENCFF821WNV", #    H3K36me3-human                lung
  # "ENCFF655YLP", #    H3K36me3-human                lung
  # "ENCFF036EBD", #     H3K4me1-human                lung
  # "ENCFF204IAN", #     H3K4me1-human                lung
  # "ENCFF262ZVW", #     H3K4me1-human                lung
  "ENCFF896NVM", #     H3K4me1-human                lung
  "ENCFF926QND", #     H3K4me1-human                lung
  "ENCFF062ZIV", #     H3K4me3-human                lung
  # "ENCFF485XPN", #     H3K4me3-human                lung
  # "ENCFF999LEM", #     H3K4me3-human                lung
  "ENCFF434JKJ", #     H3K4me3-human                lung
  # "ENCFF645EXF", #      H3K9ac-human                lung
  # "ENCFF046SLU", #      H3K9ac-human                lung
  # "ENCFF308TWI", #     H3K9me3-human                lung
  # "ENCFF305ZXD", #     H3K9me3-human                lung
  # "ENCFF275KCQ", #     H3K9me3-human                lung
  "ENCFF970KWF", #     H3K9me3-human                lung
  NULL
)
sapply(blacklist, function(id) {print(paste0("rm -Rf ", id, "*"))})
db = db[!rownames(db) %in% blacklist,]
db$file = do.call(rbind, strsplit(db$File.download.URL, "/"))[,7]
apply(db, 1, function(row) {print(paste0("ln -s ", row[["file"]], " ", row[["Biosample.term.name"]], "_", row[["Experiment.target"]], ".bw"))})
# stop("EFN")
# download
df = db
directory_prefix = "datashare/encode/"
if (!exists("DOWNLOAD")) {
  print("DOWNLOAD is not defined. To download, set DOWNLOAD=TRUE")
} else {
  ALLISGOOD = FALSE
  while (!ALLISGOOD) {
    ALLISGOOD = TRUE
      foo = apply(df, 1, function(row) {
        url = row[["File.download.URL"]]
        file = rev(strsplit(url, "/")[[1]])[1]
        file = paste0(directory_prefix, file)
        if (file.exists(file)) {
          print(paste0("File ", file, " exists."))
        } else {
          cmd = "wget"
          args = paste0("--directory-prefix ", directory_prefix, " ", url)
          print(paste(cmd, args))
          system2(cmd, args)    
        }  
      })


    foo = apply(df, 1, function(row) {
      url = row[["File.download.URL"]]
      file = rev(strsplit(url, "/")[[1]])[1]
      file = paste0(directory_prefix, file)
      md5 = tools::md5sum(file)
      if (md5 != row[["md5sum"]]) {
        print(paste0("md5 is different for ", file))
        unlink(file)
        ALLISGOOD <<- FALSE
      } else {
        print(paste0("md5 is ok for ", file))
      }  
    })
  }
}


get_bed_cgi = function(tmp_idx) {
  cgi = pfepichg38[tmp_idx,]$Islands_Name #trueIsland
  cgi = cgi[cgi!=""]
  bed3 = data.frame(do.call(rbind, strsplit(unique(cgi), ":|-")))
  names(bed3) = c("#chr", "start", "stop")
  bed3[,2] = as.numeric(bed3[,2])
  bed3[,3] = as.numeric(bed3[,3])
  bed = bed3
  bed$name = paste0(bed[,1], ":", bed[,2], "-", bed[,3])
  bed$len = bed$stop - bed$start
  bed$strand = "+"
  layout(matrix(1:3,1), respect=TRUE)
  barplot(sort(table(cgi)), las=2)
  plot(density(bed$len))
  plot(bed$len, table(cgi)[bed$name])
  return(bed)
}

write_bed_tss = function(bed, tag) {
  # build and write bed file
  bed = bed[,1:3]
  bed[,2] = round((bed[,2] + bed[,3])/2)
  head(bed)
  bed[,3] = bed[,2]+1
  bed$probes = "."
  bed$pval = 0
  bed$strand = "+"
  colnames(bed) = c("#chrom", "start", "end", "probes", "pval", "strand")
  bed[,4] = rownames(bed)
  head(bed)

  # write peaks for deeptools
  bed[,1] = as.character(bed[,1])
  bed = bed[order(bed[,1], bed[,2]),]
  options(scipen = 99)
  bedfile = paste0(tag, ".bed")
  write.table(bed, file=bedfile, sep="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
  return(bedfile)
}

get_mf_clust = function(bwfiles, bed_files, samplesLabel, prefix="all", othargs="") {
  cmd = "computeMatrix"
  args = paste0(
    "reference-point -p 32 -bs 100 -b 5000 -a 5000 -R ", paste0(bed_files, collapse=" " ), " -o mt_", prefix, ".txt.gz -S ", 
    paste0(directory_prefix, bwfiles, collapse=" ")
  ) 
  print(paste(cmd, args))
  system2(cmd, args)

  cmd = "plotHeatmap"
  args = paste0("-m mt_", prefix, ".txt.gz  --samplesLabel ", samplesLabel, " -o hm_", prefix, ".png --outFileSortedRegions hm_", prefix, ".bed ", othargs)
  print(paste(cmd, args))
  system2(cmd, args)
  mf = read.table(gzfile(paste0("mt_", prefix, ".txt.gz")), sep="\t", comment="@")
  return(mf)
}
if (!exists("mget_mf_clust")) {mget_mf_clust = memoise::memoise(get_mf_clust)}
```
## Chromatine profile

```{r hm, fig.width=9, fig.height=9}
# stop("EFN")

# idx_up = probes[["bonferroni01"      ]] 
idx_all = probes[["bonferroni01"       ]] 
idx_gain = probes[["methylation_gain"  ]] 
idx_loss = probes[["methylation_loss"  ]] 
idx_losshypo = probes[["methylation_losshypo"]] 
idx_losshyper = probes[["methylation_losshyper"]] 

# idx_up = probes[["methylation_gain"]]

# idx_isl = intersect(idx_up, rownames(pfepichg38)[pfepichg38$Relation_to_Island%in% c("Island")]) 
pfepichg38$Start_hg38 = as.numeric(pfepichg38$Start_hg38)
pfepichg38 = pfepichg38[!is.na(pfepichg38$CHR_hg38),]
set.seed(1)
idx_rnd_island  = sample(rownames(pfepichg38)[pfepichg38$Relation_to_Island%in% c("Island")], 5000) 
idx_rnd_opensea = sample(rownames(pfepichg38)[pfepichg38$Relation_to_Island%in% c("OpenSea")], 5000) 

pfepichg38$Start_hg38 = as.numeric(pfepichg38$Start_hg38)
pfepichg38 = pfepichg38[!is.na(pfepichg38$CHR_hg38),]
bed_all        = mget_fat_feats(pfepichg38[idx_all       ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
bed_gain        = mget_fat_feats(pfepichg38[idx_gain       ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
bed_loss        = mget_fat_feats(pfepichg38[idx_loss       ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
bed_losshypo      = mget_fat_feats(pfepichg38[idx_losshypo     ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
bed_losshyper      = mget_fat_feats(pfepichg38[idx_losshyper     ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
bed_rnd_island  = mget_fat_feats(pfepichg38[idx_rnd_island ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
bed_rnd_opensea = mget_fat_feats(pfepichg38[idx_rnd_opensea,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)

bed_files = c(
  write_bed_tss(bed_all       , "all"       ), 
  # write_bed_tss(bed_gain       , "gain"       ), 
  # write_bed_tss(bed_loss       , "loss"       ), 
  # write_bed_tss(bed_losshypo     , "losshypo"       ), 
  # write_bed_tss(bed_losshyper     , "losshyper"       ), 
  # write_bed_tss(bed_rnd_island , "rnd_island" ), 
  # write_bed_tss(bed_rnd_opensea, "rnd_opensea"), 
  NULL
)

ff = rbind(bed_all, bed_rnd_island, bed_rnd_opensea)
# ff = rbind(bed_all)
# ff = rbind(bed_gain)
# ff = rbind(bed_loss)

bwfiles = db$file
samplesLabel = gsub("-human", "", paste(paste0(db$Experiment.target, "_", db$Biosample.term.name), collapse=" "))
nbclust = 7
prefix = paste0("allclust", nbclust)
prefix = paste0("allclust")
mf = get_mf_clust(bwfiles, bed_files, samplesLabel, prefix=prefix, othargs=paste0("--kmeans ", nbclust, " --clusterUsingSamples 1 2 3 4 5 6 7"))


# data = as.matrix(mf[,-(1:6)])
# head(data)[,1:10]
# data = data[!apply(is.na(data), 1, any),]
# dim(data)
# pca = prcomp(data, scale=FALSE)
# v = pca$sdev * pca$sdev
# p = v / sum(v) * 100

# library(mclust)
# em = Mclust(pca$x[,1:2], G=2)
# col=em$class

# # layout(matrix(c(3,3,1,5,5,5,
# #                 2,2,4,5,5,5,
# #                 2,2,4,5,5,5,
# #                 NULL), 3, byrow=TRUE), respect=TRUE)
# layout(matrix(1:12, 3, byrow=TRUE), respect=TRUE)
# barplot(p[1:20], main="% of expl. var.")
# barplot(cumsum(p[1:20]), main="cum. % of expl. var.")
# i=1
# for (i in 1:10) {
#   j=i+1
#   plot(pca$x[,i], pca$x[,j], xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"), col=1, pch=".")
# }


clusters = read.table(paste0("hm_", prefix, ".bed"))

bed_files = sapply(unique(clusters$V13), function(tag) {
  locus_bed = clusters[clusters$V13==tag, 1:6]
  rownames(locus_bed) = locus_bed[,4]
  locus_bed_file = write_bed_tss(locus_bed, tag)
  locus_bed_file
})



# tag = "cluster_3"
# locus_bed = clusters[clusters$V13==tag, 1:6]
# rownames(locus_bed) = locus_bed[,4]
# dim(locus_bed)
# head(locus_bed)
# pbs1 = unique(unlist(ff[locus_bed[,4],]$probes))
# locus_bed_file = write_bed_tss(locus_bed, tag)
# foo = read.table(locus_bed_file)
# head(foo)
# idx = foo[,4]
# pbs2 = unique(unlist(ff[idx,]$probes))
# sum(!pbs2 %in% pbs2)


plot_terrain = function() {
  layout(matrix(c(2,2,2,1,
                2,2,2,1,
                2,2,2,1,
                3,4,5,6,
                7,8,9,10), 5, byrow=TRUE), respect=TRUE)
  par(mar=c(5.1, 4.1, 4.1, 2.1))
  idx = rownames(ewas)[ewas$pvbonf<0.01]
  xlab = "age effect (meth. prop. per year)"
  ylab = "mean methylation proportion"
  x = ewas[idx, ]$beta
  y = ewas[idx, ]$mean_beta
  z = MASS::kde2d(x, y, n = 100, lims=c(-0.002, 0.002, 0, 1))
  main = paste0("Bonferroni 1% (", length(idx), " probes)")
  fc(z, main=main, xlab=xlab, ylab=ylab, LEGEND=TRUE)
  for (bed_file in bed_files) {
    par(mar=c(2, 2.5, 3.1, 2.5))
    # idx = rownames(ewas)[ewas$pvbonf<0.01]
    # idx = probes[["cluster_1.bed"]]
    foo = read.table(bed_file)
    idx = unique(unlist(ff[read.table(bed_file)[,4],]$probes))
    x = ewas[idx, ]$beta
    y = ewas[idx, ]$mean_beta
    z = MASS::kde2d(x, y, n = 100, lims=c(-0.002, 0.002, 0, 1))
    main=paste0(bed_file, " (", length(idx), " probes)")
    fc(z, main=main, plot.axes=FALSE, color.palette=function(n) hcl.colors(n,  palette="Grays", rev=TRUE))# , levels=fc$levels, zlim=fc$zlim)
    axis(1, at=c(-0.002, 0, 0.002), labels=c(-0.002, 0, 0.002))
    axis(2, at=c(0, .5, 1), labels=c(0, .5, 1))

  }
  par(mar=c(5.1, 4.1, 4.1, 2.1))
}
plot_terrain()

```

![hm_all](hm_allclust.png)


```{r hm2, fig.width=9, fig.height=9}
metaclusters = list(c("cluster_1", "cluster_5"), c("cluster_2"), c("cluster_3", "cluster_4"), c("cluster_6"), c("cluster_7"))


bed_files = sapply(metaclusters, function(tag) {
  locus_bed = clusters[clusters$V13%in%tag, 1:6]
  rownames(locus_bed) = locus_bed[,4]
  locus_bed_file = write_bed_tss(locus_bed, paste0("metacluster", paste0(substr(tag, 9, 9), collapse="")))
  locus_bed_file
})


bed_files = c(
  bed_files,
  # write_bed_tss(bed_gain       , "gain"       ), 
  # # write_bed_tss(bed_loss       , "loss"       ), 
  # write_bed_tss(bed_losshypo     , "losshypo"       ), 
  # write_bed_tss(bed_losshyper     , "losshyper"       ), 
  write_bed_tss(bed_rnd_island , "rnd_island" ), 
  write_bed_tss(bed_rnd_opensea, "rnd_opensea"), 
  NULL
)

samplesLabel = gsub("-human", "", paste(paste0(db$Experiment.target, "_", db$Biosample.term.name), collapse=" "))
mf = get_mf_clust(bwfiles, bed_files, samplesLabel, prefix="allclustmanual")

plot_terrain()

```




![hm_all](hm_allclustmanual.png)




```{r eval=TRUE}
genome = "hg38"
if (!exists("mannotatr_build_annotations")) {mannotatr_build_annotations = memoise::memoise(annotatr::build_annotations)}
annotations = mannotatr_build_annotations(genome=genome, annotations=paste0(genome, c("_basicgenes")))

genes = lapply(bed_files, function(bed_file) {  
  tmp_regions = annotatr::read_regions(con=bed_file, genome=genome, format="bed")
  # Intersect the regions we read in with the annotations
  tmp_regions_annotated = annotatr::annotate_regions(
      regions = tmp_regions,
      annotations = annotations,
      ignore.strand = TRUE,
      quiet = FALSE)
  # A GRanges object is returned
  tmp_regions_annotated = data.frame(tmp_regions_annotated)
  dim(tmp_regions_annotated)
  # dedup
  tmp_regions_annotated$annot.type = factor(tmp_regions_annotated$annot.type, levels=paste0(genome, c("_genes_promoters", "_genes_1to5kb", "_genes_5UTRs", "_genes_exons", "_genes_introns", "_genes_3UTRs")))
  table(tmp_regions_annotated$annot.type)
  table(tmp_regions_annotated$name)
  tag = "hg38_genes_promoters"
  genes = unique(tmp_regions_annotated[tmp_regions_annotated$annot.type%in%tag & !is.na(tmp_regions_annotated$annot.symbol),]$annot.symbol)
  write.table(genes, paste0(tag, "_", bed_file, ".txt"), quote=FALSE, row.names=FALSE, col.names=FALSE)
  return(genes)
})
names(genes) = bed_files
```



```{r eval=FALSE}
# stop("EFN")
clusters = read.table(paste0("hm_", prefix, ".bed"))
k = "cluster_1"
results = clusters[clusters$V13==k, 1:6]

# annotatr
genome = "hg38"
tmp_results_filename = paste0("results_filename_", k, ".bed")
print(tmp_results_filename)
write.table(results[,1:6], file=tmp_results_filename, sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
tmp_regions = annotatr::read_regions(con=tmp_results_filename, genome=genome, format="bed")
# Build the annotations (a single GRanges object)
if (!exists("annotations_list")) {
  annotations_list = list()
}
if (! genome %in% names(annotations_list)) {
  # annots = paste0(genome, c("_lncrna_gencode"))
  # annots = paste0(genome, c('_cpgs', '_basicgenes', "_lncrna_gencode", "_enhancers_fantom"))
  # annots = paste0(genome, c('_cpgs', '_basicgenes', "_enhancers_fantom"))
  # annots = paste0(genome, c('_cpgs', '_basicgenes'))
  annots = paste0(genome, c("_basicgenes"))
  annotations_list[[genome]] = annotatr::build_annotations(genome=genome, annotations=annots)
}
annotations = annotations_list[[genome]]
# Intersect the regions we read in with the annotations
tmp_regions_annotated = annotatr::annotate_regions(
    regions = tmp_regions,
    annotations = annotations,
    ignore.strand = TRUE,
    quiet = FALSE)
# A GRanges object is returned
tmp_regions_annotated = data.frame(tmp_regions_annotated)
# dedup
tmp_regions_annotated$annot.type = factor(tmp_regions_annotated$annot.type, levels=paste0(genome, c("_genes_promoters", "_genes_1to5kb", "_genes_5UTRs", "_genes_exons", "_genes_introns", "_genes_3UTRs")))
table(tmp_regions_annotated$annot.type)
table(tmp_regions_annotated$name)
tmp_regions_annotated = tmp_regions_annotated[order(tmp_regions_annotated$name, tmp_regions_annotated$annot.type),]
tmp_regions_annotated = tmp_regions_annotated[!duplicated(paste(tmp_regions_annotated$name, tmp_regions_annotated$annot.symbol)),]
dim(tmp_regions_annotated)
table(tmp_regions_annotated$annot.type)
# aggregat by fat_feat
fat_feat_annotations = lapply(unique(tmp_regions_annotated$name), function(n) {
  # n = "chr7:2661785-2662169"
  foo = tmp_regions_annotated[tmp_regions_annotated$name==n,]
  if (sum(tmp_regions_annotated$name==n) == 1) {
    return(foo)
  } else {
    foo[["annot.symbol"]] = paste0(foo[["annot.symbol"]],collapse=";")
    foo[["annot.type"]] = paste0(foo[["annot.type"]],collapse=";")
    foo[["annot.tx_id"]] = paste0(foo[["annot.tx_id"]],collapse=";")
    return(foo[1,])
  }
})
fat_feat_annotations = do.call(rbind, fat_feat_annotations)
rownames(fat_feat_annotations) = fat_feat_annotations$name
head(fat_feat_annotations)

write.table(unique(unlist(strsplit(fat_feat_annotations$annot.symbol, ";"))), paste0(k, ".txt"), quote=FALSE, row.names=FALSE, col.names=FALSE)

# rownames(fat_feat_annotations) %in% rownames(results)
# rownames(results) %in% rownames(fat_feat_annotations)
# # inject into results
# results$annot.type   = NA
# results$annot.symbol = NA
# results$annot.tx_id  = NA
# for (idx in rownames(fat_feat_annotations)) {
#   results[idx,"annot.type"  ] = as.character(fat_feat_annotations[idx,"annot.type"  ])
#   results[idx,"annot.symbol"] = as.character(fat_feat_annotations[idx,"annot.symbol"])
#   results[idx,"annot.tx_id" ] = as.character(fat_feat_annotations[idx,"annot.tx_id" ])
# }

# results$nb_probes = NA
# results$probes = NA


# for (idx in rownames(results)) {
#   if (nrow(results)>1 & !is.matrix(dmr_indexed_probes)) {
#     idx_probes = dmr_indexed_probes[[idx]]
#   } else {
#     idx_probes = dmr_indexed_probes[,idx]
#   }
#   results[idx, "nb_probes"] = length(idx_probes)
#   results[idx, "probes"] = paste(idx_probes, collapse=";")
# }

# tmpidx = order(results[,"z_sidak_p"])
# results = results[tmpidx,]

# # results$url = sapply(results[,4], function(i){
# #   output_pdf_filename = paste0("global_results/", prefix, "_", i, ".pdf")
# #   paste0("http://epimed.univ-grenoble-alpes.fr/downloads/florent/expedition_5300/results/meth/", output_pdf_filename)
# # })

# results_filename = paste0("global_results_", prefix, ".xlsx")
# print(results_filename)
# WriteXLS::WriteXLS(results, results_filename, FreezeCol=6, FreezeRow=1, BoldHeaderRow=TRUE, AdjWidth=TRUE)





```


# Session Information

```{r, results="verbatim"}
sessionInfo()
```

