---
title: "Multi-omic integration using mediation"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---



```{r echo=FALSE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
source("common.R")
```

# Data

```{r data}
# load data
srna = mreadRDS("~/projects/gtex/results/study_gtex_trscr608.rds")
all(rownames(srna$exp_grp) == colnames(srna$data))
rownames(srna$exp_grp) = paste0(srna$exp_grp$subject, "_", srna$exp_grp$tissue)
colnames(srna$data) = rownames(srna$exp_grp)
srna$exp_grp = srna$exp_grp[order(srna$exp_grp$age, srna$exp_grp$subject),]
srna$data = srna$data[,rownames(srna$exp_grp)]

smet = mreadRDS("~/projects/gtex/results/study_gtex_meth608.rds")
all(rownames(smet$exp_grp) == colnames(smet$data))
rownames(smet$exp_grp) = paste0(smet$exp_grp$subject, "_", smet$exp_grp$tissue)
colnames(smet$data) = rownames(smet$exp_grp)
smet$exp_grp = smet$exp_grp[order(smet$exp_grp$age, smet$exp_grp$subject),]
smet$data = smet$data[,rownames(smet$exp_grp)]

genome = "hg19"
```



# Epic V1 hg38 annotated platform

```{r hg38}
library(IlluminaHumanMethylationEPICmanifest)
library(IlluminaHumanMethylationEPICanno.ilm10b5.hg38)
library(minfi)
annoEPIC_hg38 = getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b5.hg38)
all(rownames(smet$platform) %in% rownames(annoEPIC_hg38))
smet$platform[,1] = annoEPIC_hg38[rownames(smet$platform),"CHR_hg38"]
smet$platform[,2] = as.numeric(annoEPIC_hg38[rownames(smet$platform),"Start_hg38"])
smet$platform = smet$platform[!is.na(smet$platform[2]),]
smet$platform = smet$platform[!duplicated(paste0(smet$platform[,1], ":", smet$platform[,2])),]
smet$data = smet$data[rownames(smet$platform),]
genome = "hg38"

g = mreadRDS(paste0("~/projects/genes/bed_", genome, "_epimeddb.rds"))
g$tss = as.numeric(apply(g, 1, function(row) {ifelse(row[[6]]=="+", row[[2]], row[[3]])}))
head(g)
sum(rownames(g) %in% rownames(srna$data))
srna$data = srna$data[rownames(srna$data) %in% rownames(g),]
srna$platform = g[rownames(srna$data),] 

dim(srna$platform)
head(srna$platform)

dim(smet$platform)
head(smet$platform)[,1:10]
```

# Crossing genes and probes

```{r cross}
cross2bed = function(tss, probescope) {
  tss_bed_file = paste0("tss.", genome, ".bed")
  options(scipen = 999)
  write.table(tss, file=tss_bed_file, sep="\t", quote=FALSE, row.names=FALSE, col.names=FALSE)
  annots_bed_files = c(
      tss_bed_file,
      NULL    
    )
  annots = sapply(gsub(paste0(".", genome, ".bed"), "", annots_bed_files, fixed=TRUE), function(tag) {
      print(tag)
      annotatr::read_annotations(con=paste0(tag, ".", genome, ".bed"), genome=genome, name=tag, format="bed")    
      paste0("", genome, "_custom_", tag)
    })
  annotations_tss = annotatr::build_annotations(genome=genome, annotations=annots)


  tmp_probescope_filename = "tmp_probescope.bed"
  print(tmp_probescope_filename)
  options(scipen = 999)
  write.table(probescope, file=tmp_probescope_filename, sep="\t", quote=FALSE, row.names=FALSE, col.names=FALSE)
  pf_regions = annotatr::read_regions(con=tmp_probescope_filename, genome=genome, format="bed")
  pf_regions = GenomicRanges::trim(pf_regions)


  probescope_annotated = annotatr::annotate_regions(
      regions = pf_regions,
      annotations = annotations_tss,
      ignore.strand = TRUE,
      quiet = FALSE)
  # A GRanges object is returned
  df_probescope_annotated = data.frame(probescope_annotated)
  df_probescope_annotated$probescope_id = paste0(df_probescope_annotated[,1], ":", df_probescope_annotated[,2], "-", df_probescope_annotated[,3])

  dim(df_probescope_annotated)
  df_probescope_annotated$gene = rownames(tss)[as.numeric(do.call(rbind, strsplit(df_probescope_annotated$annot.id, ":"))[,2])]

  # stop("EFN")
  rownames(probescope) = paste0(probescope[,1], ":", probescope[,2]+1, "-", probescope[,3])
  df_probescope_annotated$probe = probescope[df_probescope_annotated$probescope_id,4]
  sum(is.na(df_probescope_annotated$probe))
  probescope$key = paste0(probescope[,1], ":", probescope[,2]+1)
  for (i in which(is.na(df_probescope_annotated$probe))) {
    # print(df_probescope_annotated[i,]$probe)
    idx = which(probescope$key == strsplit(df_probescope_annotated[i,]$probescope_id, "-")[[1]][1])
    df_probescope_annotated[i,]$probe = probescope[idx,4]
  }
  sum(is.na(df_probescope_annotated$probe))

  probe_gene_jointure = df_probescope_annotated[,c("gene", "probe")]
  return(probe_gene_jointure)
}
if (!exists("mcross2bed")) {mcross2bed = memoise::memoise(cross2bed)}

tss = srna$platform
tss[,2] = tss$tss
tss[,3] = tss[,2]+1
tss[,4] = rownames(tss)
tss[,5] = 0
# tss[,6] = "+"
tss = tss[order(tss[,1], tss[,2]), 1:6]

wsize=200000
wsize=10000
# targeted_probes = probes_per_bed_files$bonferroni01
targeted_probes = read.table("predimeth_probes_GSE147740_99.bed")[,4]
probescope = smet$platform[intersect(rownames(smet$platform), targeted_probes),1:6]
probescope[,2] = probescope[,2] - wsize/2
probescope[,3] = probescope[,2] + wsize
probescope[,4] = rownames(probescope)
probescope[,5] = 0
probescope[,6] = "+"
probescope[,2] = round((probescope[,2]+probescope[,3])/2) - wsize/2
probescope[,3] = probescope[,2] + wsize
probescope[,2] = sapply(probescope[,2], max, 0)
probescope = probescope[order(probescope[,1], probescope[,2]), ]

probe_gene_jointure = mcross2bed(tss, probescope)
```


Genes are assigned to probes that lie within `r wsize` bp of their genomic position, i.e. within `r wsize` bp upstream of the gene’s start or downstream of its end.”


```{r echo=TRUE, results="verbatim"}
head(probe_gene_jointure)
dim(probe_gene_jointure)
layout(matrix(1:2,1, byrow=FALSE), respect=TRUE)
barplot(table(table(probe_gene_jointure$gene)), las=2, main="#probes per gene distr.")
barplot(table(table(probe_gene_jointure$probe)), las=2, main="#genes per probe distr.")
```

# Standardization

```{r standardization}
# Filtering genes and probes
tmp_idxg = intersect(rownames(srna$data), unique(probe_gene_jointure$gene))
tmp_idxp = intersect(rownames(smet$data), unique(probe_gene_jointure$probe))
srna$data = srna$data[tmp_idxg,]
smet$data = smet$data[tmp_idxp,]
srna$platform = srna$platform[rownames(srna$data),]
smet$platform = smet$platform[rownames(smet$data),]
dim(srna$data)
dim(smet$data)

layout(matrix(1:2,1, byrow=FALSE), respect=TRUE)
plot(density(srna$data))
plot(density(smet$data))
srna$data = (srna$data - apply(srna$data, 1, mean)) / apply(srna$data, 1, sd)
smet$data = log10(smet$data / (1-smet$data))
smet$data = (smet$data - apply(smet$data, 1, mean)) / apply(smet$data, 1, sd)

srna$data = srna$data[!apply(is.na(srna$data), 1, all),]
srna$platform = srna$platform[rownames(srna$data),]
smet$data = smet$data[!apply(is.na(smet$data), 1, all),]
smet$platform = smet$platform[rownames(smet$data),]
probe_gene_jointure = probe_gene_jointure[probe_gene_jointure$gene  %in% rownames(srna$data),]
probe_gene_jointure = probe_gene_jointure[probe_gene_jointure$probe %in% rownames(smet$data),]
layout(matrix(1:2,1, byrow=FALSE), respect=TRUE)
plot(density(srna$data))
plot(density(smet$data))

```



# Multiomic dataset

GTEx / GSE213478 (methylome + transciptome)




```{r results="verbatim"}
table(srna$exp_grp$sex, srna$exp_grp$age)
# table(srna$exp_grp$subject, srna$exp_grp$age)
# stop("EFN")
```



```{r results="verbatim"}
table(srna$exp_grp$sex, srna$exp_grp$age)

srna$exp_grp = srna$exp_grp[srna$exp_grp$tissue%in%c("Lung"),]
srna$data = srna$data[,rownames(srna$exp_grp)]
dim(srna$data)

## Filtering tissue (meth)
smet$exp_grp = smet$exp_grp[smet$exp_grp$tissue%in%c("Lung"),]
smet$data = smet$data[,rownames(smet$exp_grp)]
dim(smet$data)
```



# Mediation

## Design

```{r results="verbatim"}
table(srna$exp_grp$age)
```

## Direct model 

```{r echo=TRUE}
f = expr ~ age
```

```{r m, eval=TRUE, message=FALSE, warning=FALSE}
options(scipen = 0)
d = srna$exp_grp
if (!all(colnames(srna$data) == rownames(srna$exp_grp))) {stop("colnames(srna$data) DO NOT mathch rownames(srna$exp_grp)")}
ewas_func = function(cln, f, d) {
  # d$expr = srna$data[1,]
  d$expr = cln
  m = lm(f, d)
  # boxplot(expr~age, d, las=2)
  res = c(beta=m$coefficients[[2]], pv=anova(m)[1,5])
  return(res)
}
da = epimedtools::monitored_apply(srna$data, 1, ewas_func, f=f, d=d)
da = data.frame(t(da))
```

```{r}
da[order(da$pv), ]
da$padj = p.adjust(da$pv, method="BH")
da$padj_bon = p.adjust(da$pv, method="bonferroni")

layout(matrix(1:2,1, byrow=FALSE), respect=TRUE)
smoothScatter(da$beta, -log10(da$pv))
abline(h=-log10(0.05), col=2)
smoothScatter(da$beta, -log10(da$padj))
abline(h=-log10(0.05), col=2)

sum(da$padj<0.01, na.rm=TRUE)
sum(da$padj_bon<0.05, na.rm=TRUE)

# idx_genesoi = rownames(da)[da$padj<0.01 & !is.na(da$padj)]
idx_genesoi = rownames(da)[!is.na(da$padj)]
# idx_genesoi = rownames(da[order(da$pv), ])[1:1000]
length(idx_genesoi)
```

## Indirect model 1

```{r echo=TRUE}
f1 = meth ~ age
```


```{r m1, eval=TRUE}
sub_probe_gene_jointure = probe_gene_jointure
sub_probe_gene_jointure = sub_probe_gene_jointure[sub_probe_gene_jointure[,1] %in% idx_genesoi,]
idx_probes = unique(sub_probe_gene_jointure[,2])
length(idx_probes)
d = smet$exp_grp
head(smet$exp_grp)

if (!all(colnames(smet$data) == rownames(smet$exp_grp))) {stop("colnames(smet$data) DO NOT match rownames(smet$exp_grp)")}
ewas1_func = function(cln, f1, d) {
  # d$meth = smet$data[1,]
  d$meth = cln
  m = lm(f1, d)
  # boxplot(meth~age, d, las=2)
  res = c(beta=m$coefficients[[2]], pv=anova(m)[1,5])
  return(res)
}
da1 = epimedtools::monitored_apply(smet$data[idx_probes,], 1, ewas1_func, f1=f1, d=d)
da1 = data.frame(t(da1))
```
```{r plot1}
da1$padj = p.adjust(da1$pv, method="BH")
da1$padj_bon = p.adjust(da1$pv, method="bonferroni")

layout(matrix(1:2,1, byrow=FALSE), respect=TRUE)
smoothScatter(da1$beta, -log10(da1$pv))
abline(h=-log10(0.05), col=2)
smoothScatter(da1$beta, -log10(da1$padj))
abline(h=-log10(0.05), col=2)

head(da1)

sum(da1$padj<0.01, na.rm=TRUE)
sum(da1$padj_bon<0.05, na.rm=TRUE)

idx_probesoi = rownames(da1)[da1$padj<0.01 & !is.na(da1$padj)]
# idx_probesoi = rownames(da1[order(da1$pv), ])[1:1000]

length(idx_probesoi)
```

## Indirect model 2

```{r echo=TRUE}
f2 = expr ~ meth + age
```

```{r m2, eval=TRUE, message=FALSE, warning=FALSE}
# sub_probe_gene_jointure = probe_gene_jointure[probe_gene_jointure[,2] %in% idx_probesoi,]
# sub_probe_gene_jointure = probe_gene_jointure[probe_gene_jointure[,1] %in% idx_genesoi & probe_gene_jointure[,2] %in% idx_probesoi,]
dim(sub_probe_gene_jointure)

if (!all(colnames(smet$data) == rownames(smet$exp_grp))) {stop("colnames(smet$data) DO NOT match rownames(smet$exp_grp)")}
rownames(smet$exp_grp) = paste0(smet$exp_grp$subject, "_", smet$exp_grp$age)
colnames(smet$data) = rownames(smet$exp_grp)
dim(smet$exp_grp)

if (!all(colnames(srna$data) == rownames(srna$exp_grp))) {stop("colnames(srna$data) DO NOT match rownames(srna$exp_grp)")}
rownames(srna$exp_grp) = paste0(srna$exp_grp$subject, "_", srna$exp_grp$age)
colnames(srna$data) = rownames(srna$exp_grp)
dim(srna$exp_grp)

# stop("EFN")

idx_samples = rownames(smet$exp_grp)
smet$exp_grp = smet$exp_grp[idx_samples,]
srna$exp_grp = srna$exp_grp[idx_samples,]
smet$data = smet$data[,idx_samples]
srna$data = srna$data[,idx_samples]

d = smet$exp_grp
if (!all(colnames(smet$data) == rownames(smet$exp_grp))) {stop("colnames(smet$data) DO NOT match rownames(smet$exp_grp)")}
if (!all(colnames(smet$data) == colnames(srna$data))) {stop("colnames(smet$data) DO NOT match rownames(smet$exp_grp)")}
ewas2_func = function(p, f2, d, sub_probe_gene_jointure) {
# da2 = epimedtools::monitored_apply(smet$data, 1, function(row) {
  # d$meth = smet$data[1,]
  # p = idx_probesoi[1]
  # p = idx_probesoi[40]
  sub_sub_probe_gene_jointure = sub_probe_gene_jointure[sub_probe_gene_jointure[,2]==p,]
  res = lapply(sub_sub_probe_gene_jointure[,1], function(g, p, f2, d) {
    # # g = sub_sub_probe_gene_jointure[1,1]
    # # print(g)
    d$meth = smet$data[p,]
    d$expr = srna$data[g,]
    m = lm(f2, d)
    s = summary(m)
    # plot(d$meth, d$expr)
    # abline(m)
    res = data.frame(g=g, p=p, beta_meth=m$coefficients[[2]], beta_age=m$coefficients[[3]], pv_meth=s$coefficients[2,4], pv_age=s$coefficients[3,4])
  }, p=p, f2=f2, d=d)
  res = do.call(rbind, res)
  return(res)
}
da2 = epimedtools::monitored_apply(t(t(idx_probesoi)), 1, ewas2_func, f2=f2, d=d, sub_probe_gene_jointure=sub_probe_gene_jointure)
da2 = do.call(rbind, da2)
rownames(da2) = paste0(da2$g, "_", da2$p)
# # check
# p = "cg11667847" ; g = "TMEM203"
# meth = smet$data[p,]
# expr = srna$data[g,]
# age = smet$exp_grp$age
# lm(expr ~ meth + age)
# m = lm(f2, d)
# m 
# da2[paste0(g, "_", p),]
```

```{r plot2, eval=TRUE}
layout(matrix(1:2,1, byrow=FALSE), respect=TRUE)
smoothScatter(da2$beta_meth, -log10(da2$pv_meth))
smoothScatter(da2$beta_age, -log10(da2$pv_age))
```

## Results

```{r mediation}
rownames(sub_probe_gene_jointure) = paste0(sub_probe_gene_jointure$gene, "_", sub_probe_gene_jointure$probe)
results = sub_probe_gene_jointure

results$a    = da1[results$probe,]$beta
results$b    = da2[rownames(results),]$beta_meth
results$ade  = da2[rownames(results),]$beta_age
results$c    = da[results$gene,]$beta
results$acme = results$a * results$b
results$pmed = results$acme / results$c
results$chr = srna$platform[results$gene,]$chr
results$tss = srna$platform[results$gene,]$tss
results$pos = smet$platform[results$probe,]$pos
results$delta = abs(results$pos - results$tss)
results$pv_tot = da[results$gene,]$pv
results$pv1 = da1[results$probe,]$pv
results$pv_age = da2[rownames(results),]$pv_age
results$pv_meth = da2[rownames(results),]$pv_meth
results$pv1 = da1[results$probe,]$pv
results$pmax2 = apply(results[, c("pv_meth", "pv1")], 1, max)^2
results[is.na(results$pmax2),]$pmax2 = 1
results$pmax2fdr = p.adjust(results$pmax2, method = "BH")
saveRDS(results, paste0("results_", nrow(results), ".rds"))

tmp_idxg = 1:length(unique(results$gene))
names(tmp_idxg) = unique(results$gene)
results$idxg = tmp_idxg[results$gene]

tmp_idxp = 1:length(unique(results$probe))
names(tmp_idxp) = unique(results$probe)
results$idxp = tmp_idxp[results$probe]


# ntop = 20
# tmp_idx = rownames(results[order(results$pmax2),])[1:ntop]
tmp_idx = rownames(results[
  results$pmax2<0.05 & 
  results$pv_tot<0.05 & 
  # results$pv_age<0.05 & 
  sign(results$ade) == sign(results$acme) & 
  abs(results$c)>0.00000001,])
length(tmp_idx)

layout(matrix(1:2, 1, byrow=FALSE), respect=TRUE)

plot(results$c, -log10(results$pv_tot), pch=".", xlab="TOTAL", ylab="-log10(pval)")
points(results[tmp_idx,]$c, -log10(results[tmp_idx,]$pv_tot), col=2)

plot(results$acme, -log10(results$pmax2), pch=".", xlab="ACME", ylab="-log10(pmax2)")
points(results[tmp_idx,]$acme, -log10(results[tmp_idx,]$pmax2), col=2)

# plot(results$idxg, results$idxp, pch=".", xlab="genes", ylab="probes")
# abline(a=0, b=length(unique(results$probe))/length(unique(results$gene)), col="grey")
# points(results[tmp_idx,]$idxg, results[tmp_idx,]$idxp, col=adjustcolor(2, alpha.f=30/ntop), pch=16)

plot(results$acme, results$ade, pch=".")
points(results[tmp_idx,]$acme, results[tmp_idx,]$ade, col=2)
abline(h=0, v=0, col=1)

plot(results$acme, results$delta, pch=".")
points(results[tmp_idx,]$acme, results[tmp_idx,]$delta, col=2)

```





```{r eval=FALSE}
names(predimeth_probes_per_bed_files)
sapply(predimeth_probes_per_bed_files, length)

results$cluster = 1
for (k in names(predimeth_probes_per_bed_files)[1:5]) {
  results[results$probe %in% predimeth_probes_per_bed_files[[k]],]$cluster = k
}
table(results$cluster, useNA="always")

layout(matrix(1:6, 2, byrow=FALSE), respect=TRUE)

plot(results$acme, -log10(results$pmax2fdr), pch=".", xlab="ACME", ylab="-log10 pmax2 + FDR 5% (BH)", col=as.numeric(as.factor(results$cluster)))
# ntop = 20
# tmp_idx = rownames(results[order(results$pmax2),])[1:ntop]
tmp_idx = rownames(results[results$pmax2fdr<0.05 ,])
points(results[tmp_idx,]$acme, -log10(results[tmp_idx,]$pmax2fdr), col=as.numeric(as.factor(results[tmp_idx,]$cluster)))

plot(results$idxg, results$idxp, pch=".", xlab="genes", ylab="probes", col=as.numeric(as.factor(results$cluster)))
abline(a=0, b=length(unique(results$probe))/length(unique(results$gene)), col="grey")
points(results[tmp_idx,]$idxg, results[tmp_idx,]$idxp, col=adjustcolor(as.numeric(as.factor(results[tmp_idx,]$cluster)), alpha.f=30/ntop), pch=16)

plot(results$acme, results$ade, pch=".", col=as.numeric(as.factor(results$cluster)))
points(results[tmp_idx,]$acme, results[tmp_idx,]$ade, col=as.numeric(as.factor(results[tmp_idx,]$cluster)))
abline(h=0, v=0, col=1)

plot(results$acme, results$delta, pch=".", col=as.numeric(as.factor(results$cluster)))
points(results[tmp_idx,]$acme, results[tmp_idx,]$delta, col=as.numeric(as.factor(results[tmp_idx,]$cluster)))


subresults = results[tmp_idx,]
subresults$score1 = 0
subresults$score2 = 0
foo = sort(table(subresults[tmp_idx,]$gene))
subresults$score1 = foo[subresults$gene]
bar = sort(table(subresults[tmp_idx,]$probe))
subresults$score2 = bar[subresults$probe]
subresults$score3 = subresults$score1 + subresults$score2
subresults[order(subresults$score3),]

```{r resulst xlsx, results="verbatim", echo=TRUE}
head(results[tmp_idx,])
dim(results[tmp_idx,])
table_mediation_filename = paste0("table_mediation_", gse, ".xlsx")
table_mediation = results[tmp_idx,]
WriteXLS::WriteXLS(table_mediation, table_mediation_filename)
```

Table of mediation results is exported as [`r table_mediation_filename`](`r table_mediation_filename`).

```{r mediation_pkg, eval=TRUE}
for (i in tmp_idx) {
  layout(matrix(1:6,2, byrow=FALSE), respect=TRUE)

  plot(results$acme, results$delta, pch=".")
  points(results[tmp_idx,]$acme, results[tmp_idx,]$delta, col=2)
  points(results[i,]$acme, results[i,]$delta, col=4, pch=16, cex=3)
  text(results[i,]$acme, results[i,]$delta, i)

  plot(results$acme, results$ade, pch=".")
  points(results[tmp_idx,]$acme, results[tmp_idx,]$ade, col=2)
  points(results[i,]$acme, results[i,]$ade, col=4, pch=16, cex=3)
  text(results[i,]$acme, results[i,]$ade, i)
  abline(h=0, v=0, col=1)

  p = results[i,"probe"]
  g = results[i,"gene"]
  d = smet$exp_grp
  d$meth = smet$data[p,]
  d$expr = srna$data[g,]

  f = expr ~ age
  f1 = meth ~ age
  f2 = expr ~ meth + age
  fit = lm(f, d)
  fita = lm(f1, d)
  fitb = lm(f2, d)
  if (!exists("mmediate")) {mmediate = memoise::memoise(mediation::mediate)}
  fitmed = mmediate(fita, fitb, treat ="age", mediator = "meth")
  summary(fitmed)
  results[i,]$ade

  main = paste0(g, " " , p)
  plot(fitmed, main=paste0("Mediation ", main))
  plot(as.factor(as.character(d$age)), d$expr, las=2, xlab="", ylab="expression (z-score)")
  plot(as.factor(as.character(d$age)), d$meth, las=2, xlab="", ylab="methylation (M-value, z-score)")
  plot(d$meth, d$expr, las=2, xlab="methylation (M-value, z-score)", ylab="expression (z-score)", col=as.numeric(as.factor(as.character(d$age))))
}
```


```{r mediation_tcga, eval=TRUE}
# key = "TCGA-LUAD"
key = "TCGA-LUSC"
smetkc = mreadRDS(paste0("~/projects/datashare/", key, "/study_preproc_", key, ".rds"))
e = smetkc$exp_grp
idx = intersect(e[e$tissue_status == "normal",]$id_patient, e[e$tissue_status == "tumoral",]$id_patient)
e = e[e$id_patient%in%idx,]
tab1 = table(e[e$tissue_status == "normal",]$id_patient)
tab2 = table(e[e$tissue_status == "tumoral",]$id_patient)
idx = intersect(names(tab1)[tab1==1], names(tab2)[tab2==1])    
e = e[e$id_patient%in%idx,]
tab = table(e$id_patient, e$tissue_status)
tab
emetkc = e
smetkc$data = smetkc$data[,rownames(emetkc)]
smetkc$exp_grp = smetkc$exp_grp[rownames(emetkc),]

srnakc = mreadRDS(paste0("~/projects/tcga_studies/study_", key, "_trscr.rds"))
e = srnakc$exp_grp
idx = intersect(e[e$tissue_status == "normal",]$id_patient, e[e$tissue_status == "tumoral",]$id_patient)
e = e[e$id_patient%in%idx,]
tab1 = table(e[e$tissue_status == "normal",]$id_patient)
tab2 = table(e[e$tissue_status == "tumoral",]$id_patient)
idx = intersect(names(tab1)[tab1==1], names(tab2)[tab2==1])    
e = e[e$id_patient%in%idx,]
tab = table(e$id_patient, e$tissue_status)
tab
ernakc = e
srnakc$data = srnakc$data[,rownames(ernakc)]
srnakc$exp_grp = srnakc$exp_grp[rownames(ernakc),]



for (i in tmp_idx[results[tmp_idx,]$probe %in% rownames(smetkc$data) & results[tmp_idx,]$gene %in% rownames(srnakc$data)]) {
  layout(matrix(1:8,2, byrow=FALSE), respect=TRUE)

  plot(results$acme, results$delta, pch=".")
  points(results[tmp_idx,]$acme, results[tmp_idx,]$delta, col=2)
  points(results[i,]$acme, results[i,]$delta, col=4, pch=16, cex=3)
  text(results[i,]$acme, results[i,]$delta, i)

  plot(results$acme, results$ade, pch=".")
  points(results[tmp_idx,]$acme, results[tmp_idx,]$ade, col=2)
  points(results[i,]$acme, results[i,]$ade, col=4, pch=16, cex=3)
  text(results[i,]$acme, results[i,]$ade, i)
  abline(h=0, v=0, col=1)

  p = results[i,"probe"]
  g = results[i,"gene"]
  d = smet$exp_grp
  d$meth = smet$data[p,]
  d$expr = srna$data[g,]

  f = expr ~ age
  f1 = meth ~ age
  f2 = expr ~ meth + age
  fit = lm(f, d)
  fita = lm(f1, d)
  fitb = lm(f2, d)
  if (!exists("mmediate")) {mmediate = memoise::memoise(mediation::mediate)}
  fitmed = mmediate(fita, fitb, treat ="age", mediator = "meth")
  summary(fitmed)
  results[i,]$ade

  main = paste0(g, " " , p)
  plot(fitmed, main=paste0("Mediation ", main))
  plot(d$meth, d$expr, las=2, xlab="methylation (M-value, z-score)", ylab="expression (z-score)", col=as.numeric(as.factor(as.character(d$age))))
  plot(as.factor(as.character(d$age)), d$meth, las=2, xlab="", ylab="methylation (M-value, z-score)")
  plot(as.factor(as.character(d$age)), d$expr, las=2, xlab="", ylab="expression (z-score)")


  dmetkc = smetkc$exp_grp
  dmetkc$meth = smetkc$data[p,]
  drnakc = srnakc$exp_grp
  drnakc$expr = srnakc$data[g,]

  plot(as.factor(dmetkc$tissue_status), dmetkc$meth, las=2, xlab="", ylab="methylation (beta-value)")
  plot(as.factor(drnakc$tissue_status), drnakc$expr, las=2, xlab="", ylab="expression (pseudolog norm count)")


}
```







```{r check, eval=FALSE}
d = smet$exp_grp
if (!all(colnames(smet$data) == rownames(smet$exp_grp))) {stop("colnames(smet$data) DO NOT match rownames(smet$exp_grp)")}
if (!all(colnames(smet$data) == colnames(srna$data))) {stop("colnames(smet$data) DO NOT match rownames(smet$exp_grp)")}
library(mediation)
for (i in 1:10) {
  # p = "cg14164839" ; g = "RPS6KA1"
  p = results[i,"probe"]
  g = results[i,"gene"]
  d$meth = smet$data[p,]
  d$expr = srna$data[g,]

  f = f = expr ~ age
  f1 = meth ~ age
  f2 = expr ~ meth + age
  fit = lm(f, d)
  fita = lm(f1, d)
  fitb = lm(f2, d)
  fitmed = mediate(fita, fitb, treat="age", mediator = "meth")
  print(summary(fitmed))
  print("ADE")
  print(fitmed$z0)
  print(fitb$coefficients[[3]])
  print(results[paste0(g, "_", p), ]$ade)  
  print("ACME")
  print(fitmed$d0)
  print(fita$coef[[2]] * fitb$coef[[2]])
  print(results[paste0(g, "_", p), ]$acme)  
}
```


# combp

```{r combp, eval=FALSE}
head(results)
genes = unique(results$gene)
gidxed_res = epimedtools::monitored_apply(t(t(genes)), 1, function(g) {
  r = results[results$gene==g,]
})
names(gidxed_res) = genes
# for (pval_thresh in c(0.000001) { #, 0.00001, 0.0001) {
empty_res = data.frame(X.chrom=NA, start=NA, end=NA, min_p=NA, n_probes=NA, z_p=NA, z_sidak_p=NA, gene=NA)[-1,]
pval_thresh = 0.000001
  foo = epimedtools::monitored_apply(t(t(genes)), 1, function(g) {
    print(g)
    # r = gidxed_res[[1]]
    r = gidxed_res[[g]]
    if (nrow(r) == 1 ) {
      return(empty_res)
    }
    prefix = paste0(g, "_", wsize, "bp")
    # build and write bed file
    bed = r[,c("chr", "pos", "pos", "probe", "pmax2")]
    # head(bed)
    bed[,3] = bed[,2]+1
    bed$probe = rownames(r)
    bed$strand = "+"
    colnames(bed) = c("chrom", "start", "end", "probes", "pval", "strand")
    # head(bed)
    # write ewas res for combp
    bed_ewas_filename = paste0("ewas4combp_", prefix,".bed")
    bed[,1] = as.character(bed[,1])
    bed = bed[order(bed[,1], bed[,2]),]
    options(scipen=999)
    write.table(bed,file=bed_ewas_filename , sep="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
    options(scipen=0)
    # bed_on_disk = read.table(bed_ewas_filename, sep="\t", header=TRUE)

    # Run comb-p, run!!
    prefix2 = paste0(prefix, "_", pval_thresh)
    combp_res_acf_file = paste0("dmr_", prefix2, ".acf.txt")
    if (file.exists(combp_res_acf_file)) {
      combp_res_acf = read.table(combp_res_acf_file, comment="w", header=TRUE)
      if (nrow(combp_res_acf)==0) {
        return(empty_res)
      }
    }
    combp_res_probes_file = paste0("dmr_", prefix2, ".fdr.bed.gz")    
    if (!file.exists(combp_res_probes_file)) {
      Sys.setenv(PYTHONPATH = "/summer/epistorage/opt/combined-pvalues/") # export PYTHONPATH="/summer/epistorage/opt/combined-pvalues/"
      cmd = "/summer/epistorage/opt/combined-pvalues/cpv/comb-p"
      arg = paste0("pipeline -c 5 --seed ", pval_thresh, " --dist 1000 -p dmr_", prefix2," --region-filter-p 0.05 --region-filter-n 2 ", bed_ewas_filename)
      print(paste(cmd, arg))
      system2(cmd, arg)
    }

    # recuparation du fichier dmrbycombp1000_SGCE.fdr.bed après comb-p
    # ewas = read.table(paste0("ewas4combp_", prefix, ".bed"), header=TRUE)
    # combp_res_probes = read.table(gzfile(paste0("dmr_", prefix2, ".fdr.bed.gz")), comment="@", header=TRUE)
    combp_res_region_filename = paste0("dmr_", prefix2, ".regions-t.bed")
    if (file.exists(combp_res_region_filename)) {  
      combp_res_region = read.table(combp_res_region_filename, comment="@", header=TRUE)
      # head(combp_res_region[order(-combp_res_region$n_probes, combp_res_region$z_sidak_p),])
      # reg = combp_res_region[order(-combp_res_region$n_probes, combp_res_region$z_sidak_p),][1,]
      # probes = ewas[as.character(ewas[,1])==as.character(reg[[1]]) & ewas[,2]>=reg[[2]] & ewas[,2]<=reg[[3]],4]
      if (nrow(combp_res_region)!=0) {
        combp_res_region$gene = g
      } else {
        return(empty_res)
      }
    } else {
      return(empty_res)
    }
    return(combp_res_region)
  })
  foo = do.call(rbind,foo)

  # build and write bed file
  bed = foo
  head(bed)
  bed[,4] = bed$gene
  bed[,6] = "+"
  bed = bed[,1:6]
  colnames(bed) = c("#chrom", "start", "end", "probes", "pval", "strand")
  head(bed)
  # write ewas res for combp
  bed_dmrs_filename = paste0("dmrs_", wsize,".bed")
  bed[,1] = as.character(bed[,1])
  bed = bed[order(bed[,1], bed[,2]),]
  options(scipen=999)
  write.table(bed,file=bed_dmrs_filename , sep="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
  options(scipen=0)

```


# Session Information

```{r, results="verbatim"}
sessionInfo()
```




