---
title: "Post processing"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---


```{r echo=FALSE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
source("common.R")
info = list(start_time = Sys.time())
```

```{r params}
# # source("params_default.R")
if (!exists("gse"))             gse = "GSE147740"  ;
if (!exists("model_formula"))   model_formula = "meth~age"  ;
if (!exists("model_func_name")) model_func_name = "modelcalllm"         ;
newas = "1000000"
neighb = 1000
runmax = 45
# if (file.exists(paste0(model_func_name, ".R"))) {
#   source(paste0(model_func_name, ".R"))
# }
# study_filename = paste0("./datashare/", gse, "/study_preproc_", gse, ".rds")

# if (!exists("newas"))           newas = "1000000"                            ;
# if (!exists("neighb"))          neighb = "1000"                           ;

# bed_ewas_filename = paste0("ewas4combp_", gse, "_", model_func_name, "_", model_formula, ".bed")
# rds_ewas_filename = paste0("ewas_", gse, "_", model_func_name, "_", model_formula, ".rds")
study_filename = paste0("~/projects/datashare/", gse, "/study_preproc_", gse, "_", model_func_name, "_", model_formula, "_ewas", newas, "_nn", neighb, ".rds")

rds_ewas_filename = paste0("ewas_", gse, "_", model_func_name, "_", model_formula, ".rds")
```

```{r loading study and ewas}
s = mreadRDS(study_filename)
# Loading ewas
ewas = mreadRDS(rds_ewas_filename)
# PATCH to not compute this each times
if (!"pvbonf" %in% colnames(ewas)) {
  ewas = data.frame(ewas)
  ewas$pvbonf = p.adjust(ewas[,"pv"], method="bonferroni")
  ewas$mean_beta = apply(s$data[rownames(ewas),], 1, mean)
  ewas$sd_beta = apply(s$data[rownames(ewas),], 1, sd)
  saveRDS(ewas, rds_ewas_filename)
}
```

# Post proc iterative model

```{r postproc itertive model}
print(gse)
print(newas)
res = lapply(0:runmax, function(run) {
  print(run)
  res = data.frame(gse=gse, newas=newas, run=run)
  info_file = paste0("info_model_r", run, "_", gse, "_modelcalllm_meth~age_ewas", newas, "_nn", neighb, ".rds")
  models_file = paste0(  "models_r", run, "_", gse, "_modelcalllm_meth~age_ewas", newas, "_nn", neighb, ".rds")
  if (file.exists(info_file)) {
    res$done=TRUE
    # print(info_file)
    # info_file = "info_model_r0_GSE40279_modelcalllm_meth~age_ewas500000_nn1000.rds"
    infomodel = readRDS(info_file)
    # models_file = "models_r0_GSE40279_modelcalllm_meth~age_ewas500000_nn1000.rds"
    models = readRDS(models_file)
    res$exec_time = as.numeric(infomodel$exec_time)
    res$rmse = infomodel$gender.bootstrap.RMSE
    res$nbprobes = infomodel$gender.bootstrap.nb_probes_mod
    res$probes = list(models[[2]]$coeff$probes)
  } else {
    print(paste0(info_file, " missing"))
    res$done=FALSE
    res$exec_time = 0.1
    res$rmse = 0
    res$nbprobes = 0
    res$probes = list()
  }
  res
})
res = do.call(rbind, res)
# res = data.frame(lapply(data.frame(res, stringsAsFactors=FALSE), unlist), stringsAsFactors=FALSE)
print(dim(res))
res

# export probes as bed file
lastrun = runmax
print(gse)
sub_res = res[res$gse==gse & res$run<=lastrun & res$newas==newas, ]
foo = apply(sub_res, 1, function(row) {
  data.frame(run=row$run, probes=unlist(row$probes))
})
foo = do.call(rbind, foo)
head(foo)
dim(foo)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
pf450k = data.frame(getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19))
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
pfepic = data.frame(getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19))
if (all(foo$probes %in% rownames(pf450k))) {
  pforig = pf450k
} else {
  pforig = pfepic  
}   
bed = cbind(pforig[foo$probes,1:4], run=foo$run)
bed = cbind(bed, strand=bed[,3])
bed[,3] = bed[,2]+1
bed$version="hg19"
colnames(bed) = c("#chrom", "start", "end", "name", "score", "strand", "version")
head(bed)

# write ewas res for combp
bed[,1] = as.character(bed[,1])
bed = bed[order(bed[,1], bed[,2]),]
bed_filename = paste0("predimeth_probes_", gse, "_", lastrun, ".bed")
bed_predimeth_hg19 = bed
write.table(bed_predimeth_hg19, file=bed_filename , sep="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
```

# hg38

```{r hg38}
library(IlluminaHumanMethylationEPICmanifest)
library(IlluminaHumanMethylationEPICanno.ilm10b5.hg38)
library(minfi)
pfepichg38 = getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b5.hg38)

table(pfepichg38$Relation_to_Island, useNA="always")
pfepichg38[pfepichg38$Relation_to_Island == "N_Shelf", ]$Relation_to_Island = "Shelf" 
pfepichg38[pfepichg38$Relation_to_Island == "N_Shore", ]$Relation_to_Island = "Shore" 
pfepichg38[pfepichg38$Relation_to_Island == "S_Shelf", ]$Relation_to_Island = "Shelf" 
pfepichg38[pfepichg38$Relation_to_Island == "S_Shore", ]$Relation_to_Island = "Shore" 
table(pfepichg38$Relation_to_Island, useNA="always")
pfepichg38$Relation_to_Island = factor(pfepichg38$Relation_to_Island, levels = c("Island", "Shore", "Shelf", "OpenSea"))






colpals_cpgdensity=c("chartreuse4", "darkorange", "aquamarine3", "darkblue")

colpals_cpgdensity=c("#228B22", "#F4A460", "#87CEFA", "#191970")
colpals_cpgdensity=c("#006400", "#E69F00", "#56B4E9", "#00007F")
names(colpals_cpgdensity) = levels(pfepichg38$Relation_to_Island)


# library("IlluminaHumanMethylationEPICv2anno.20a1.hg38")
# pfepichg38 = data.frame(getAnnotation(IlluminaHumanMethylationEPICv2anno.20a1.hg38))
# dim(pfepichg38)
# pfepichg38 = pfepichg38[pfepichg38$EPICv1_Loci %in% rownames(ewas),]
# dim(pfepichg38)
# pfepichg38 = pfepichg38[!duplicated(pfepichg38$EPICv1_Loci),]
# rownames(pfepichg38) = pfepichg38$EPICv1_Loci
# pfepichg38$Relation_to_Island = factor(pfepichg38$Relation_to_Island, levels = c("Island", "Shore", "Shelf", "OpenSea"))
# dim(pfepichg38)



all(rownames(pfepichg38) %in% rownames(ewas))
if (!all(rownames(ewas) %in% rownames(pfepichg38))) {
  if (nrow(ewas) < 500000) {PF450K = TRUE}
  if (!exists("PF450K")) {stop("problem annotation hg38.")}
  ewas = ewas[intersect(rownames(ewas), rownames(pfepichg38)),]
  ewas = ewas[!is.na(pfepichg38[rownames(ewas), "CHR_hg38"]),]
  pfepichg38 = pfepichg38[rownames(ewas),]
  ewas$Relation_to_Island = pfepichg38[rownames(ewas),]$Relation_to_Island
  rm("PF450K")
} else {
  pfepichg38 = pfepichg38[rownames(ewas),]
  ewas$Relation_to_Island = pfepichg38[rownames(ewas),]$Relation_to_Island
}
```

```{r echo=TRUE, results="verbatim"}
sum(pfepichg38$Relation_to_Island=="Island") / nrow(pfepichg38)
foo = do.call(rbind, strsplit(sort(names(table(pfepichg38$Islands_Name)))[-1], ":|-"))
foo = data.frame(foo)
foo$len = as.numeric(foo[,3]) - as.numeric(foo[,2])
sum(foo$len) / nrow(foo)

chrom_len = read.table("~/projects/datashare/genomes/Homo_sapiens/UCSC/hg38/hg38.chrom.sizes")
sum(foo$len) / sum(chrom_len[1:25,2])

```





# Inject episignature in ewas df

```{r}
# stop("EFN")
litterature_models = readRDS("litterature_models.rds")
probes_bonf01 = rownames(ewas)[ewas$pvbonf<0.01]
probes_hannum = rownames(litterature_models$hannum_model_mc$coeff)
probes_horvath = rownames(litterature_models$horvath_model_mc$coeff)
probes_phenoage = read.table("aging-v10i4-101414-supplementary-material-SD2.csv", sep=",", header=TRUE)$CpG[-1]
probes_predimeth = bed_predimeth_hg19[,4]
predimeth_it = bed_predimeth_hg19[,5]
names(predimeth_it) = rownames(bed_predimeth_hg19)


ewas$gene = pfepichg38[rownames(ewas),]$UCSC_RefGene_Name
ewas$group = pfepichg38[rownames(ewas),]$UCSC_RefGene_Group

tmp_names = colnames(ewas)

# 6 first colums as bed
ewas$chr_hg38 = pfepichg38[rownames(ewas),]$CHR_hg38
ewas$start_hg38 = pfepichg38[rownames(ewas),]$Start_hg38
ewas$end_hg38 = pfepichg38[rownames(ewas),]$End_hg38
ewas$name = rownames(ewas)
ewas$beta_age = ewas$beta
ewas$strand_hg38 = pfepichg38[rownames(ewas),]$Strand_hg38

# signature
ewas$bonf01 = 0
ewas[intersect(probes_bonf01, rownames(ewas)),]$bonf01 = 1
ewas$predimeth = 0
ewas[intersect(probes_predimeth, rownames(ewas)),]$predimeth = 1
ewas$predimeth_it = NA
if (!all(names(predimeth_it) %in% rownames(ewas))) {stop("not all names(predimeth_it) in rownames(ewas)")}
ewas[names(predimeth_it),]$predimeth_it = predimeth_it
ewas$hannum = 0
ewas[intersect(probes_hannum, rownames(ewas)),]$hannum = 1
ewas$horvath = 0
ewas[intersect(probes_horvath, rownames(ewas)),]$horvath = 1
ewas$phenoage = 0
ewas[intersect(probes_phenoage, rownames(ewas)),]$phenoage = 1
```





# Stat Desc

```{r stat_desc, fig.width=9, fig.height=6}
s = mreadRDS(study_filename)
exp_grp = s$exp_grp

d = density(apply(s$data, 1, mean), col="grey", lty=3)
d_m = density(apply(s$data[,rownames(exp_grp)[exp_grp$gender=="M"]], 1, mean), bw=d$bw)
d_f = density(apply(s$data[,rownames(exp_grp)[exp_grp$gender=="F"]], 1, mean), bw=d$bw)

idx_Islan = rownames(pfepichg38)[pfepichg38$Relation_to_Island=="Island" ]
idx_Shore = rownames(pfepichg38)[pfepichg38$Relation_to_Island=="Shore"  ]
idx_Shelf = rownames(pfepichg38)[pfepichg38$Relation_to_Island=="Shelf"  ]
idx_OpenS = rownames(pfepichg38)[pfepichg38$Relation_to_Island=="OpenSea"]

d_Islan = density(apply(s$data[idx_Islan,], 1, mean), bw=d$bw)
d_Shore = density(apply(s$data[idx_Shore,], 1, mean), bw=d$bw)
d_Shelf = density(apply(s$data[idx_Shelf,], 1, mean), bw=d$bw)
d_OpenS = density(apply(s$data[idx_OpenS,], 1, mean), bw=d$bw)


exp_grp$qtage = cut(exp_grp$age, unique(quantile(exp_grp$age, 0:40/40)), include.lowest=TRUE)
table(exp_grp$qtage, useNA="ifany")
d_yng = density(apply(s$data[,rownames(exp_grp)[exp_grp$qtage==levels(exp_grp$qtage)[1]]], 1, mean), bw=d$bw)
d_old = density(apply(s$data[,rownames(exp_grp)[exp_grp$qtage==rev(levels(exp_grp$qtage))[1]]], 1, mean), bw=d$bw)

rm("s")
```

```{r child = '05_postprocstatdesc.Rmd', eval=TRUE}
```


# EWAS

```{r volcanobf, fig.width=7, fig.height=7}
x = ewas$beta
y = ewas$lpv
xlab = "age effect (meth. prop. per year)"
ylab = "-log10(pval)"
main = paste0("methylation~age (", nrow(ewas), " probes)")
col = as.numeric(ewas$pvbonf<0.01)+ 1
layout(1, respect=TRUE)
plot(x, y, xlab=xlab, ylab=ylab, main=main, col=adjustcolor(col, alpha.f=.3), pch=16)
legend("topleft", pch=16, col=1:2, c(paste0(">= 1% (", sum(!ewas$pvbonf<0.01), " probes)"), paste0("< 1% (", sum(ewas$pvbonf<0.01), " probes)")), title="Bonferroni")
```



```{r ewas, fig.width=7, fig.height=7}
fc = function (x = seq(0, 1, length.out = nrow(z)), y = seq(0, 1, 
    length.out = ncol(z)), z, xlim = range(x, finite = TRUE), 
    ylim = range(y, finite = TRUE), zlim = range(z, finite = TRUE), 
    levels = pretty(zlim, nlevels), nlevels = 20, color.palette = function(n) hcl.colors(n, 
        "YlOrRd", rev = TRUE), col = color.palette(length(levels) - 
        1), plot.title, plot.axes, key.title, key.axes, key.border = NULL, 
    asp = NA, xaxs = "i", yaxs = "i", las = 1, axes = TRUE, frame.plot = axes, LEGEND=FALSE,
    ...) 
{
    if (missing(z)) {
        if (!missing(x)) {
            if (is.list(x)) {
                z <- x$z
                y <- x$y
                x <- x$x
            }
            else {
                z <- x
                x <- seq.int(0, 1, length.out = nrow(z))
            }
        }
        else stop("no 'z' matrix specified")
    }
    else if (is.list(x)) {
        y <- x$y
        x <- x$x
    }
    if (any(diff(x) <= 0) || any(diff(y) <= 0)) 
        stop("increasing 'x' and 'y' values expected")
    if (LEGEND) { 
      mar.orig <- (par.orig <- par(c("mar", "las", "mfrow")))$mar
      # on.exit(par(par.orig))
      w <- (3 + mar.orig[2L]) * par("csi") * 2.54
      # layout(matrix(c(2, 1), ncol = 2L), widths = c(1, lcm(w)))
      # par(las = las)
      # mar <- mar.orig
      # mar[4L] <- mar[2L]
      # mar[2L] <- 1
      # par(mar = mar)
      par(mar=c(5.1, 5.1, 4.1, 5.1))

      plot.new()
      plot.window(xlim = c(0, 1), ylim = range(levels), xaxs = "i", 
          yaxs = "i")
      rect(0, levels[-length(levels)], 1, levels[-1L], col = col, 
          border = key.border)
      if (missing(key.axes)) {
          if (axes) 
              axis(4)
      }
      else key.axes
      box()
      if (!missing(key.title)) 
          key.title
      mar <- mar.orig
      # mar[4L] <- 1
      par(mar = mar)
    }
    plot.new()
    plot.window(xlim, ylim, "", xaxs = xaxs, yaxs = yaxs, asp = asp)
    .filled.contour(x, y, z, levels, col)
    if (missing(plot.axes)) {
        if (axes) {
            title(main = "", xlab = "", ylab = "")
            Axis(x, side = 1)
            Axis(y, side = 2)
        }
    }
    else plot.axes
    if (frame.plot) 
        box()
    if (missing(plot.title)) 
        title(...)
    else plot.title
    return(list(levels=levels, zlim=zlim))
}


  layout(matrix(c(2,2,2,1,
                  2,2,2,1,
                  2,2,2,1,
                  3,4,5,6), 4, byrow=TRUE), respect=TRUE)
  par(mar=c(5.1, 4.1, 4.1, 2.1))
  idx = rownames(ewas)[ewas$pvbonf<0.01]
  xlab = "age effect (meth. prop. per year)"
  ylab = "mean methylation proportion"
  x = ewas[idx, ]$beta
  y = ewas[idx, ]$mean_beta
  z = MASS::kde2d(x, y, n = 100, lims=c(-0.002, 0.002, 0, 1))
  main = paste0("Bonferroni 1% (", length(idx), " probes)")
  fc(z, main=main, xlab=xlab, ylab=ylab, LEGEND=TRUE)
  # nlevels = 20
  # zlim = range(z$z, finite = TRUE)
  # levels = pretty(zlim, nlevels)
  # x <- z$x
  # y <- z$y
  # grid <- expand.grid(x=x, y=y)
  # grid$z <- as.vector(z$z)
  # contourplot(z ~ x * y, grid, cuts = 20,labels=FALSE, main=main)
  # levelplot(z ~ x * y, grid, cuts = 10, colorkey=FALSE, main=main)
  colpals=c("Greens", "Grays", "Purples", "Blues")
  names(colpals) = levels(ewas$Relation_to_Island)
  for (k in levels(ewas$Relation_to_Island)) {
    par(mar=c(2, 2.5, 3.1, 2.5))
    idx = rownames(ewas)[ewas$pvbonf<0.01]
    idx = intersect(idx, rownames(ewas)[ewas$Relation_to_Island%in% k])
    x = ewas[idx, ]$beta
    y = ewas[idx, ]$mean_beta
    z = MASS::kde2d(x, y, n = 100, lims=c(-0.002, 0.002, 0, 1))
    main=paste0(k, " (", length(idx), " probes)")
    fc(z, main=main, plot.axes=FALSE, color.palette=function(n) hcl.colors(n,  palette=colpals[k], rev=TRUE))# , levels=fc$levels, zlim=fc$zlim)
    axis(1, at=c(-0.002, 0, 0.002), labels=c(-0.002, 0, 0.002))
    axis(2, at=c(0, .5, 1), labels=c(0, .5, 1))

  }
  par(mar=c(5.1, 4.1, 4.1, 2.1))

# x = ewas[idx, ]$beta
# y = ewas[idx, ]$mean_beta
# z = MASS::kde2d(x, y, n = 100, lims=c(-0.002, 0.002, 0, 1))

#      levelplot(z ~ x * y, grid), cuts = 50, ), xlab="",
#                ylab="", main="Weird Function", sub="with log scales",
#                colorkey = FALSE, region = TRUE)



```







```{r child = '05_postprocewas.Rmd', eval=TRUE}
```




# missMethyl


## Figure 

```{r missmethyl}
options(scipen = 0)
library(missMethyl)
# params
array.type = "EPIC"
plot.bias = FALSE
prior.prob = TRUE
anno = NULL
equiv.cpg = TRUE
fract.counts = TRUE
sig.genes = FALSE
# collection
# kegg = missMethyl:::.getKEGG()
go  = missMethyl:::.getGO()
head(go$idTable)
tmp_idTable = go$idTable
dim(tmp_idTable)
tmp_idTable = tmp_idTable[tmp_idTable$ONTOLOGY=="BP",]
dim(tmp_idTable)
collection = go$idList[tmp_idTable$GOID]
# reference and target
sig.cpg = rownames(ewas)[ewas$pvbonf<0.01]
all.cpg = rownames(ewas)
genomic.features = "ALL" #, "TSS200", "TSS1500", "Body",         "1stExon", "3'UTR", "5'UTR", "ExonBnd"), sig.genes = FALSE)
# Go!
if (!exists("mmissmethyl_gsameth")) {mmissmethyl_gsameth=memoise::memoise(missMethyl::gsameth)}
result <- mmissmethyl_gsameth(sig.cpg = sig.cpg, all.cpg = all.cpg, 
    collection = collection, array.type = array.type, 
    plot.bias = plot.bias, prior.prob = prior.prob, anno = anno, 
    equiv.cpg = equiv.cpg, fract.counts = fract.counts, 
    genomic.features = genomic.features, sig.genes = sig.genes)
result <- merge(tmp_idTable, result, by.x = "GOID", 
    by.y = "row.names")
rownames(result) <- result$PathwayID
result$TERM_short = substr(result$TERM, 1, 35)
result$EXP = result$N * length(sig.cpg) / length(all.cpg) 
head(result[order(result$P.DE),-(1:3)], 30)







# library(enrichplot)
# library(clusterProfiler)

# # Supposez que res est déjà trié et filtré
# ego <- head(result[order(result$P.DE),-(1:3)], 30)
# ego$ID <- ego$TERM  # Rename if needed
# ego$Description <- ego$TERM
# ego$GeneRatio <- ego$DE / ego$N
# ego$Count = ego$N
# ego$p.adjust = signif(ego$FDR,3)

# # enrichplot::dotplot(ego, x = "GeneRatio", showCategory = 10)
# enrichplot:::dotplot.enrichResult(ego, x = "GeneRatio", showCategory = 10)


missmethylfig = function (result) {
  # Supposons que 'res' est la sortie de la fonction gometh
  # res <- gometh(sig.cpg = sig.cpgs, all.cpg = all.cpgs, collection = "GO", array.type = "450K")

  # Sélection des 20 premiers termes les plus significatifs
  ntop=20
  top_res <- result[order(result$P.DE),][1:ntop,]
  top_res <- top_res[order(top_res$N),]

  layout(matrix(1:2, 1), respect=TRUE) ;
  # Préparation du graphique
  par(mar = c(5.1, 15, 4.1, 0.1))  # marges pour laisser de la place aux labels
  bp = barplot(
    top_res$N, 
    # -log10(top_res$P.DE), 
    names.arg = top_res$TERM,
    horiz = TRUE,
    las = 1,  # rotation des noms pour qu'ils soient horizontaux
    col = "lightgrey", 
    xlab = "#genes",
    main = paste0("Top ", ntop, " GO terms (BP)"),  
    border=0
  )
  barplot(top_res$DE,  horiz = TRUE,col = "darkgrey", add=TRUE, border=0, xaxt="n", yaxt="n")
  barplot(top_res$EXP, horiz = TRUE,col = "black",    add=TRUE, border=0, xaxt="n", yaxt="n")
  legend("bottomright", c("N", "DM", "EXP"), col=c("lightgrey", "darkgrey", "black"), pt.cex=2, pch=15)
  par(xpd = NA)
  text(top_res$N, bp, paste0("FDR=", signif(top_res$FDR, 1)), pos=4)
  par(xpd = FALSE)
  put_a_letter("a")

  par(mar=c(5.1, 4.1, 4.1, 2.1))
  par(mar=c(0,0,0,0))
  result = result[result$FDR<0.05,]
  tag_blacklist = c("cell", "of", "in", "to", "by")
  tab = sort(table(unlist(strsplit(result$TERM," "))))
  tab = tab[!names(tab) %in% tag_blacklist]
  par(xpd = NA)
  wordcloud::wordcloud(names(tab), tab, min.freq=3, use.r.layout=TRUE)
  par(xpd = FALSE)
  par(mar=c(5.1, 4.1, 4.1, 2.1))
  put_a_letter("b")
}



missmethylfig(result)
w=12
pdf(paste0("fig_", gse, "_missmethyl.pdf"), width=w, height=w/2)
missmethylfig(result)
dev.off()
```

## missMethyl per epigenomic contexts

```{r}
for (feat in c("Island", "Shore", "Shelf", "OpenSea")) {
  print(feat)
  sig.cpg = rownames(ewas)[ewas$pvbonf<0.01 & ewas$Relation_to_Island %in% feat]
  all.cpg = rownames(ewas)[ewas$Relation_to_Island %in% feat]
  genomic.features = "ALL" #, "TSS200", "TSS1500", "Body",         "1stExon", "3'UTR", "5'UTR", "ExonBnd"), sig.genes = FALSE)
  # Go!
  if (!exists("mmissmethyl_gsameth")) {mmissmethyl_gsameth=memoise::memoise(missMethyl::gsameth)}
  result <- mmissmethyl_gsameth(sig.cpg = sig.cpg, all.cpg = all.cpg, 
      collection = collection, array.type = array.type, 
      plot.bias = plot.bias, prior.prob = prior.prob, anno = anno, 
      equiv.cpg = equiv.cpg, fract.counts = fract.counts, 
      genomic.features = genomic.features, sig.genes = sig.genes)
  result <- merge(tmp_idTable, result, by.x = "GOID", 
      by.y = "row.names")
  rownames(result) <- result$PathwayID
  result$TERM_short = substr(result$TERM, 1, 35)
  result$EXP = result$N * length(sig.cpg) / length(all.cpg) 
  head(result[order(result$P.DE),-(1:3)], 30)
  missmethylfig(result)
  put_a_letter(feat)
}


```

# Pair-Correlation
```{r pair correlation}
head(ewas)
dim(ewas)
head(pfepichg38)
dim(pfepichg38)

# center_on = ""
co = "Shore"
# co = "Shelf"
idx = rownames(ewas)[ewas$pvbonf<0.01]
df_co = pfepichg38[idx, c("chr", "pos", "Relation_to_Island")]
# targeted feature
tf = "Island"
idx = rownames(ewas)
df_tf = pfepichg38[idx, c("chr", "pos", "Relation_to_Island")]

df_co = df_co[df_co$Relation_to_Island%in%co, ]
df_tf = df_tf[df_tf$Relation_to_Island%in%tf, ]

chrs = sort(unique(df_co$chr))
tmp_df_co = lapply(chrs, function(chr) { 
  # chr = chrs[1]
  print(chr)
  tmp_df_co = df_co[df_co$chr%in%chr, ]
  tmp_df_tf = df_tf[df_tf$chr%in%chr, ]
  tmp_df_co[[paste0("nearest_", tf)]] = sapply(tmp_df_co$pos, function(pos) {
    # pos = tmp_df_co$pos[1]
    tmp_delta = abs(pos - tmp_df_tf$pos)
    rownames(tmp_df_tf)[which(tmp_delta == min(tmp_delta))][1]
  })
  tmp_df_co[[paste0("dist_nearest_", tf)]] = abs(tmp_df_co$pos - tmp_df_tf[tmp_df_co[[paste0("nearest_", tf)]],]$pos)
  tmp_df_co[[paste0("dist_nearest_", tf)]]
  tmp_df_co
})
tmp_df_co = do.call(rbind, tmp_df_co)
plot(density(tmp_df_co[[paste0("dist_nearest_", tf)]]))
plot(tmp_df_co[[paste0("dist_nearest_", tf)]], ewas[rownames(tmp_df_co),]$mean_beta)


x = tmp_df_co[[paste0("dist_nearest_", tf)]]
y = ewas[rownames(tmp_df_co),]$mean_beta
# y = (ewas[rownames(tmp_df_co),]$mean_beta - ewas[tmp_df_co[[paste0("nearest_", tf)]],]$mean_beta) / 2 + 0.5

# y = ewas[tmp_df_co$nearest_Shore,]$mean_beta
# d = density(y)
xlab = paste0("distance to nearest ", tf, " (bp)")
ylab = "mean methylation proportion"
main = paste0(co, " (", length(x), " probes)")
xlim = c(0, 2500)
if (co=="Shore") xlim = c(0, 2500)
if (co=="Shelf") xlim = c(0, 5000)
ylim = c(0, 1)
custom_hm(x, y, xlim, ylim, xlab, ylab, main, col=1, LOG=TRUE, nbin=100)
  
```




# PCA methylome GTEx 




```{r loading GTEx methylome, eval=!exists("NOPCA")}
if(!exists("tissue")) {tissue = "Lung"}
s = study_gtex_meth = mreadRDS("~/projects/gtex/results/study_gtex_meth608.rds")
levels(s$exp_grp$tissue) = sub(" - .*$", "", levels(s$exp_grp$tissue))

layout(matrix(1:2, 1), respect=TRUE)
barplot(sort(table(s$exp_grp$tissue)), las=2)
```

```{r pca, eval=!exists("NOPCA")}
data = t(s$data)
head(data)[,1:10]
pca = mprcomp(data, scale=FALSE)
v = pca$sdev * pca$sdev
p = v / sum(v) * 100
names(p) = paste0("PC", 1:length(p))

pc_max = which(cumsum(p) >= 90)[1]
cols = as.numeric(s$exp_grp[rownames(pca$x),]$tissue)
# library(mclust)
# em = Mclust(pca$x[,1:2], G=2)
# col=em$class

# layout(matrix(c(3,3,1,5,5,5,
#                 2,2,4,5,5,5,
#                 2,2,4,5,5,5,
#                 NULL), 3, byrow=TRUE), respect=TRUE)
layout(matrix(1:8, 2, byrow=FALSE), respect=TRUE)
barplot(p[1:pc_max], main="% of expl. var.")
barplot(cumsum(p[1:pc_max]), main="cum. % of expl. var.")
i=1
for (i in 1:6) {
  j=i+1
  plot(pca$x[,i], pca$x[,j], xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"), col=adjustcolor(cols, alpha.f=0.3), pch=16)
  for (lev in levels(s$exp_grp$tissue)) {
    tmp_idx = rownames(s$exp_grp)[s$exp_grp$tissue==lev]
    text(mean(pca$x[tmp_idx,i]), mean(pca$x[tmp_idx,j]), lev)
  }
}
# plot.new()
# legend("topright", levels(s$exp_grp$tissue), pch=16, col=1:length(levels(s$exp_grp$tissue)))



layout(1, respect=TRUE)
foo = pca$x[,1:pc_max]
rownames(foo) = s$exp_grp[rownames(foo),]$tissue
rownames(foo)[duplicated(rownames(foo))] = NA
nodePar = list(lab.cex = 0.6, pch = c(NA, 19), 
                cex = 0.7, col = cols)
dendr = (hclust(dist(foo), method="ward.D2"))
# set(dendr, "leaves_cols", cols)

plot(dendr)

```

## Figure


```{r, eval=!exists("NOPCA")}
plot_gtexpcafig = function() {
  layout(matrix(c(
  2,2,2,2,1,1,
  2,2,2,2,1,1,
  2,2,2,2,3,3,
  2,2,2,2,3,3,
  NULL), 4, byrow=TRUE), respect=TRUE)

  barplot(p[1:pc_max], main="% of expl. var.", las=2, )
  i=1
  j=i+1
  plot(pca$x[,i], pca$x[,j], xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"), col=adjustcolor(cols, alpha.f=0.3), pch=16, main="PCA on GTEx methylome")
  for (lev in levels(s$exp_grp$tissue)) {
    tmp_idx = rownames(s$exp_grp)[s$exp_grp$tissue==lev]
    text(mean(pca$x[tmp_idx,i]), mean(pca$x[tmp_idx,j]), lev)
  }

  plot(dendr, xlab="", ylab="height (Ward D2 based)")
}

plot_gtexpcafig()
w = 8
pdf("fig_gtexpca.pdf", width=w, height=2/3*w)
# png("fig_gtexpca.png", width=1200, height=800, res=150)
plot_gtexpcafig()
dev.off()

```











# ENCODE / deeptools


## Histone marks

### Marks Associated with Transcriptional Activation:

1. **H3K27ac**  
   - Associated with active enhancers and promoters.  
   - Indicates a transcriptionally permissive environment.

2. **H3K36me3**  
   - Found in coding regions of actively transcribed genes.  
   - Linked to transcriptional elongation and genomic integrity.

3. **H3K4me1**  
   - Associated with enhancers (often primed or active).  
   - Reflects regulatory potential but does not guarantee activity.

4. **H3K4me3**  
   - Marks active or poised promoters.  
   - Strongly linked to transcriptional initiation.

5. **H3K9ac**  
   - Found at active promoters and enhancers.  
   - Associated with euchromatic, transcriptionally permissive environments.

---

### Marks Associated with Transcriptional Repression:
6. **H3K27me3**  
   - Linked to transcriptional repression, often mediated by the Polycomb Repressive Complex 2 (PRC2).  
   - Characteristic of facultative heterochromatin (regions that are repressed but potentially reactivatable).

7. **H3K9me3**  
   - A mark of constitutive heterochromatin, found in inactive genomic regions such as centromeres and telomeres.  
   - Recruits proteins like HP1 (Heterochromatin Protein 1) to maintain a condensed chromatin state.

---

### Summary:


| **Histone Mark** | **Activation/Repression**   | **Associated Regions**               |
|-------------------|-----------------------------|-------------------------------------|
| **H3K27ac**       | Activation                 | Active enhancers, promoters          |
| **H3K36me3**      | Activation                 | Coding regions                       |
| **H3K4me1**       | Activation (potential)     | Primed/active enhancers              |
| **H3K4me3**       | Activation                 | Active or poised promoters           |
| **H3K9ac**        | Activation                 | Active promoters, enhancers          |
| **H3K27me3**      | Repression                 | Facultative heterochromatin  (PRC2)  |
| **H3K9me3**       | Repression                 | Constitutive heterochromatin (HP1 )  |





```{r encode_db}
# db = read.table("~/projects/meth3d/results/entex/metadata.insituhic.tsv", sep="\t", header=TRUE)
db = read.table("encode_lung.tsv", sep="\t", header=TRUE)
# db = read.table("encode_ovary.tsv", sep="\t", header=TRUE)
rownames(db) = db[,1]
dim(db)
head(db)
table(db[,3])

table(db$File.format, db$File.type)
db = db[db$File.type == "bigWig",]

table(db$Experiment.target)
db$Experiment.target = factor(db$Experiment.target, levels=c(
  "H3K4me3-human"  , # Activation            | Active or poised promoters           |
  "H3K4me1-human"  , # Activation (potential)| Primed/active enhancers              |
  "H3K27me3-human" , # Repression            | Facultative heterochromatin  (PRC2)  |
  "H3K27ac-human"  , # Activation            | Active enhancers, promoters          |
  "H3K9ac-human"   , # Activation            | Active promoters, enhancers          |
  "H3K36me3-human" , # Activation            | Coding regions                       |
  "H3K9me3-human"  , # Repression            | Constitutive heterochromatin (HP1 )  |
  "CTCF-human", 
  NULL
))
db = db[order(db$Experiment.target),]
db = db[db$Experiment.target %in% c(
  "H3K27me3-human" , # Repression            | Facultative heterochromatin  (PRC2)  |
  "H3K4me3-human"  , # Activation            | Active or poised promoters           |
  "H3K4me1-human"  , # Activation (potential)| Primed/active enhancers              |
  "H3K27ac-human"  , # Activation            | Active enhancers, promoters          |
  # "H3K9ac-human"   , # Activation            | Active promoters, enhancers          |
  # "H3K36me3-human" , # Activation            | Coding regions                       |
  # "H3K9me3-human"  , # Repression            | Constitutive heterochromatin (HP1 )  |
  NULL
),]

blacklist = c(
  # "ENCFF331IOZ", #     H3K27ac-human                lung
  "ENCFF028TDS", #     H3K27ac-human                lung
  # "ENCFF386ZSF", #    H3K27me3-human                lung
  "ENCFF238XGJ", #    H3K27me3-human                lung
  "ENCFF440HPT", #    H3K27me3-human                lung
  # "ENCFF507BZS", #    H3K36me3-human                lung
  # "ENCFF170SDA", #    H3K36me3-human                lung
  # "ENCFF821WNV", #    H3K36me3-human                lung
  # "ENCFF655YLP", #    H3K36me3-human                lung
  "ENCFF036EBD", #     H3K4me1-human                lung
  # "ENCFF204IAN", #     H3K4me1-human                lung
  "ENCFF262ZVW", #     H3K4me1-human                lung
  "ENCFF896NVM", #     H3K4me1-human                lung
  "ENCFF926QND", #     H3K4me1-human                lung
  "ENCFF062ZIV", #     H3K4me3-human                lung
  "ENCFF485XPN", #     H3K4me3-human                lung
  # "ENCFF999LEM", #     H3K4me3-human                lung
  "ENCFF434JKJ", #     H3K4me3-human                lung
  # "ENCFF645EXF", #      H3K9ac-human                lung
  # "ENCFF046SLU", #      H3K9ac-human                lung
  # "ENCFF308TWI", #     H3K9me3-human                lung
  # "ENCFF305ZXD", #     H3K9me3-human                lung
  # "ENCFF275KCQ", #     H3K9me3-human                lung
  "ENCFF970KWF", #     H3K9me3-human                lung
  NULL
)
sapply(blacklist, function(id) {print(paste0("rm -Rf ", id, "*"))})
db = db[!rownames(db) %in% blacklist,]
db$file = do.call(rbind, strsplit(db$File.download.URL, "/"))[,7]
apply(db, 1, function(row) {print(paste0("ln -s ", row[["file"]], " ", row[["Biosample.term.name"]], "_", row[["Experiment.target"]], ".bw"))})
# stop("EFN")
# download
df = db
directory_prefix = "datashare/encode/"
if (!exists("DOWNLOAD")) {
  print("DOWNLOAD is not defined. To download, set DOWNLOAD=TRUE")
} else {
  ALLISGOOD = FALSE
  while (!ALLISGOOD) {
    ALLISGOOD = TRUE
      foo = apply(df, 1, function(row) {
        url = row[["File.download.URL"]]
        file = rev(strsplit(url, "/")[[1]])[1]
        file = paste0(directory_prefix, file)
        if (file.exists(file)) {
          print(paste0("File ", file, " exists."))
        } else {
          cmd = "wget"
          args = paste0("--directory-prefix ", directory_prefix, " ", url)
          print(paste(cmd, args))
          system2(cmd, args)    
        }  
      })


    foo = apply(df, 1, function(row) {
      url = row[["File.download.URL"]]
      file = rev(strsplit(url, "/")[[1]])[1]
      file = paste0(directory_prefix, file)
      md5 = tools::md5sum(file)
      if (md5 != row[["md5sum"]]) {
        print(paste0("md5 is different for ", file))
        unlink(file)
        ALLISGOOD <<- FALSE
      } else {
        print(paste0("md5 is ok for ", file))
      }  
    })
  }
}



write_bed_tss = function(bed, tag) {
  # build and write bed file
  bed = bed[,1:3]
  bed[,2] = round((bed[,2] + bed[,3])/2)
  head(bed)
  bed[,3] = bed[,2]+1
  bed$probes = "."
  bed$pval = 0
  bed$strand = "+"
  colnames(bed) = c("#chrom", "start", "end", "probes", "pval", "strand")
  bed[,4] = rownames(bed)
  head(bed)

  # write peaks for deeptools
  bed[,1] = as.character(bed[,1])
  bed = bed[order(bed[,1], bed[,2]),]
  options(scipen = 99)
  bedfile = paste0(tag, ".bed")
  write.table(bed, file=bedfile, sep="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
  options(scipen = 0)
  return(bedfile)
}

get_mf_clust = function(bwfiles, bed_files, samplesLabel, prefix="all", othargs="") {
  cmd = "computeMatrix"
  args = paste0(
    "reference-point -p 32 -bs 100 -b 5000 -a 5000 -R ", paste0(bed_files, collapse=" " ), " -o mt_", prefix, ".txt.gz -S ", 
    paste0(directory_prefix, bwfiles, collapse=" ")
  ) 
  print(paste(cmd, args))
  system2(cmd, args)

  cmd = "plotHeatmap"
  args = paste0("-m mt_", prefix, ".txt.gz  --samplesLabel ", samplesLabel, " -o hm_", prefix, ".png --outFileSortedRegions hm_", prefix, ".bed ", othargs)
  print(paste(cmd, args))
  system2(cmd, args)
  mf = read.table(gzfile(paste0("mt_", prefix, ".txt.gz")), sep="\t", comment="@")
  return(mf)
}
if (!exists("mget_mf_clust")) {mget_mf_clust = memoise::memoise(get_mf_clust)}
```
## Chromatine profile

### Heatmap (raw)

```{r hm, fig.width=9, fig.height=9}
# stop("EFN")
# table(ewas$predimeth, ewas$cluster, useNA="always")



probes_signature = rownames(ewas)[
  ewas$predimeth == 1    |
  ewas$hannum    == 1    |
  ewas$horvath   == 1    |
  ewas$phenoage   == 1    |
  ewas$bonf01      == 1      
]

# stop("EFN")
# idx_up = probes[["bonferroni01"      ]] 
# idx_all = probes[["bonferroni01"       ]] 
# idx_gain = probes[["methylation_gain"  ]] 
# idx_loss = probes[["methylation_loss"  ]] 
# idx_losshypo = probes[["methylation_losshypo"]] 
# idx_losshyper = probes[["methylation_losshyper"]] 

# idx_up = probes[["methylation_gain"]]

# idx_isl = intersect(idx_up, rownames(pfepichg38)[pfepichg38$Relation_to_Island%in% c("Island")]) 
pfepichg38$Start_hg38 = as.numeric(pfepichg38$Start_hg38)
pfepichg38 = pfepichg38[!is.na(pfepichg38$CHR_hg38),]
set.seed(1)
idx_rnd_island  = sample(rownames(pfepichg38)[pfepichg38$Relation_to_Island%in% c("Island")], 5000) 
idx_rnd_opensea = sample(rownames(pfepichg38)[pfepichg38$Relation_to_Island%in% c("OpenSea")], 5000) 

pfepichg38$Start_hg38 = as.numeric(pfepichg38$Start_hg38)
pfepichg38 = pfepichg38[!is.na(pfepichg38$CHR_hg38),]
bed_signature        = mget_fat_feats(pfepichg38[probes_signature       ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
# bed_bonf01        = mget_fat_feats(pfepichg38[probes_bonf01       ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
# bed_gain        = mget_fat_feats(pfepichg38[idx_gain       ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
# bed_loss        = mget_fat_feats(pfepichg38[idx_loss       ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
# bed_losshypo      = mget_fat_feats(pfepichg38[idx_losshypo     ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
# bed_losshyper      = mget_fat_feats(pfepichg38[idx_losshyper     ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
# bed_rnd_island  = mget_fat_feats(pfepichg38[idx_rnd_island ,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)
# bed_rnd_opensea = mget_fat_feats(pfepichg38[idx_rnd_opensea,], "CHR_hg38", "Start_hg38", extend_region_dist=1000)

# fat_feats= rbind(bed_bonf01, bed_rnd_island, bed_rnd_opensea)
fat_feats= rbind(bed_signature)
# fat_feats= rbind(bed_gain)
# fat_feats= rbind(bed_loss)


bed_files = c(
  write_bed_tss(bed_signature       , "all"       ), 
  # write_bed_tss(bed_bonf01       , "all"       ), 
  # write_bed_tss(bed_gain       , "gain"       ), 
  # write_bed_tss(bed_loss       , "loss"       ), 
  # write_bed_tss(bed_losshypo     , "losshypo"       ), 
  # write_bed_tss(bed_losshyper     , "losshyper"       ), 
  # write_bed_tss(bed_rnd_island , "rnd_island" ), 
  # write_bed_tss(bed_rnd_opensea, "rnd_opensea"), 
  NULL
)


bwfiles = db$file
samplesLabel = gsub("-human", "", paste(paste0(db$Experiment.target, "_", db$Biosample.term.name), collapse=" "))
nbclust = 5
prefix = paste0("allclust", nbclust)
prefix = paste0("allclust_", gse)
mf = mget_mf_clust(bwfiles, bed_files, samplesLabel, prefix=prefix, othargs=paste0("--kmeans ", nbclust)) #, " --clusterUsingSamples 1 2 3 4 5 6 7"))


# data = as.matrix(mf[,-(1:6)])
# head(data)[,1:10]
# data = data[!apply(is.na(data), 1, any),]
# dim(data)
# pca = prcomp(data, scale=FALSE)
# v = pca$sdev * pca$sdev
# p = v / sum(v) * 100

# library(mclust)
# em = Mclust(pca$x[,1:2], G=2)
# col=em$class

# # layout(matrix(c(3,3,1,5,5,5,
# #                 2,2,4,5,5,5,
# #                 2,2,4,5,5,5,
# #                 NULL), 3, byrow=TRUE), respect=TRUE)
# layout(matrix(1:12, 3, byrow=TRUE), respect=TRUE)
# barplot(p[1:20], main="% of expl. var.")
# barplot(cumsum(p[1:20]), main="cum. % of expl. var.")
# i=1
# for (i in 1:10) {
#   j=i+1
#   plot(pca$x[,i], pca$x[,j], xlab=paste0("PC", i, "(", signif(p[i], 3), "%)"), ylab=paste0("PC", j, "(", signif(p[j], 3), "%)"), col=1, pch=".")
# }


clusters = read.table(paste0("hm_", prefix, ".bed"))

bed_files = sapply(unique(clusters$V13), function(tag) {
  locus_bed = clusters[clusters$V13==tag, 1:6]
  rownames(locus_bed) = locus_bed[,4]
  locus_bed_file = write_bed_tss(locus_bed, tag)
  locus_bed_file
})



plot_terrain = function(bed_files) {
  layout(matrix(c(2,2,2,1,
                2,2,2,1,
                2,2,2,1,
                3,4,5,6,
                7,8,9,10), 5, byrow=TRUE), respect=TRUE)
  par(mar=c(5.1, 4.1, 4.1, 2.1))
  idx = rownames(ewas)[ewas$pvbonf<0.01]
  xlab = "age effect (meth. prop. per year)"
  ylab = "mean methylation proportion"
  x = ewas[idx, ]$beta
  y = ewas[idx, ]$mean_beta
  z = MASS::kde2d(x, y, n = 100, lims=c(-0.002, 0.002, 0, 1))
  main = paste0("Bonferroni 1% (", length(idx), " probes)")
  fc(z, main=main, xlab=xlab, ylab=ylab, LEGEND=TRUE)
  for (bed_file in bed_files) {
    par(mar=c(2, 2.5, 3.1, 2.5))
    # idx = rownames(ewas)[ewas$pvbonf<0.01]
    # idx = probes[["cluster_1.bed"]]
    locus_bed = read.table(bed_file)
    idx = unique(unlist(fat_feats[locus_bed[,4],]$probes))
    x = ewas[idx, ]$beta
    y = ewas[idx, ]$mean_beta
    z = MASS::kde2d(x, y, n = 100, lims=c(-0.002, 0.002, 0, 1))
    main=paste0(bed_file, " (", length(idx), " probes)")
    fc(z, main=main, plot.axes=FALSE, color.palette=function(n) hcl.colors(n,  palette="Grays", rev=TRUE))# , levels=fc$levels, zlim=fc$zlim)
    axis(1, at=c(-0.002, 0, 0.002), labels=c(-0.002, 0, 0.002))
    axis(2, at=c(0, .5, 1), labels=c(0, .5, 1))

  }
  par(mar=c(5.1, 4.1, 4.1, 2.1))
}

```

![hm_all](`r paste0("hm_", prefix, ".png")`)

### Clusters (raw)

```{r, fig.width=9, fig.height=9}
plot_terrain(bed_files)
```

```{r child = '05_postprocdeeptools.Rmd', eval=TRUE}
```


# PrediMeth model

```{r}
res = lapply(0:runmax, function(run) {
	print(run)
	res = data.frame(gse=gse, newas=newas, run=run)
	info_file = paste0("info_model_r", run, "_", gse, "_modelcalllm_meth~age_ewas", newas, "_nn", neighb, ".rds")
	models_file = paste0(   "models_r", run, "_", gse, "_modelcalllm_meth~age_ewas", newas, "_nn", neighb, ".rds")
	if (file.exists(info_file)) {
		res$done=TRUE
		# print(info_file)
		# info_file = "info_model_r0_GSE40279_modelcalllm_meth~age_ewas500000_nn1000.rds"
		infomodel = readRDS(info_file)
		# models_file = "models_r0_GSE40279_modelcalllm_meth~age_ewas500000_nn1000.rds"
		models = readRDS(models_file)
		res$exec_time = as.numeric(infomodel$exec_time)
		res$rmse = infomodel$gender.bootstrap.RMSE
		res$nbprobes = infomodel$gender.bootstrap.nb_probes_mod
		res$probes = list(models[[2]]$coeff$probes)
	} else {
		print(paste0(info_file, " missing"))
		res$done=FALSE
		res$exec_time = 0.1
		res$rmse = 0
		res$nbprobes = 0
		res$probes = list()
	}
	res
})
res = do.call(rbind, res)
print(dim(res))
```

```{r}
# probes_per_bed_files$metacluster15.bed
# unique(unlist(res$probes))
foo = sapply(probes_per_bed_files, function(pbs) {
	sum(unique(unlist(res$probes)) %in% pbs)
})
barplot(foo, las=2)
barplot(foo, las=2)

bar = lapply(1:nrow(res), function(i) {
	sapply(probes_per_bed_files, function(pbs) {
		sum(res[i,]$probes[[1]] %in% pbs)
	})
})
bar = do.call(rbind, bar)
res = cbind(res, bar)




layout(matrix(1:8, 2), respect=TRUE)
foo = sapply(names(probes_per_bed_files), function(k) {
	plot(cumsum(res[,k]), main=k, col=which(names(probes_per_bed_files)%in%k))
})

layout(1, respect=TRUE)
plot(0, 0, col=0, xlim=c(0, nrow(res)), ylim=c(0,1))
foo = sapply(names(probes_per_bed_files)[-c(6:7)], function(k) {
	points(cumsum(res[,k]) / sum(res[,k]), col=which(names(probes_per_bed_files)%in%k))
})
legend("bottomright", pch=1, col=1:length(probes_per_bed_files), paste0(names(probes_per_bed_files), " (", lapply(probes_per_bed_files, length), ")"))




head(ewas)


layout(matrix(1:8, 2), respect=TRUE)
foo = sapply(names(probes_per_bed_files), function(k) {
	pbs = probes_per_bed_files[[k]]
	plot(density(ewas[pbs,]$lpv), main=k, col=which(names(probes_per_bed_files)%in%k))
})
legend("bottomright", pch=1, col=1:length(probes_per_bed_files), paste0(names(probes_per_bed_files), " (", lapply(probes_per_bed_files, length), ")"))
```

```{r loading data}
# study_filename = paste0("./datashare/", gse, "/study_preproc_", gse, ".rds")

###### RMSE/R2 




```





```{r child = '05_postprocpredimethstats.Rmd', eval=TRUE}
```










































# GSEA

```{r}
tcga_studies = c(    
  "TCGA-LUSC" , # 40                
  "TCGA-LUAD" , # 27                
  "TCGA-KIRC" , # 160                
  "TCGA-KIRP" , # 42                
  # "TCGA-KICH" , # 0                
  # "TCGA-LGG"  , # 0                
  # "TCGA-GBM"  , # 1                
  # "TCGA-ACC"  , # 0                
  "TCGA-HNSC" , # 49                
  # "BRB"       , # 7                
  "TCGA-BRCA" , # 85                
  "TCGA-BLCA" , # 21                
  # "TCGA-CESC" , # 3                
  # "TCGA-CHOL" , # 9                
  "TCGA-COAD" , # 36                
  # "TCGA-DLBC" , # 0                
  # "TCGA-ESCA" , # 16                
  # "TCGA-LAML" , # 0                
  "TCGA-LIHC" , # 49                
  # "TCGA-MESO" , # 0                
  # "TCGA-PAAD" , # 10                
  # "TCGA-PCPG" , # 3                
  "TCGA-PRAD" , # 50                
  # "TCGA-READ" , # 7                
  # "TCGA-SARC" , # 2                
  # "TCGA-SKCM" , # 2                
  # "TCGA-STAD" , # 2                
  # "TCGA-TGCT" , # 0                
  "TCGA-THCA" , # 56                
  # "TCGA-THYM" , # 2                    
  # "TCGA-UCS"  , # 0                
  # "TCGA-UVM"  , # 0                
  # "TCGA-OV"   , # 0                     
  NULL
)

# tcga_studies = c(    
#   # "TCGA-KIRC" , # 160                
#   # "TCGA-KIRP" , # 42                
#   # # "TCGA-KICH" , # 0                
#   # # "TCGA-LGG"  , # 0                
#   # # "TCGA-GBM"  , # 1                
#   # # "TCGA-ACC"  , # 0                
#   # "TCGA-HNSC" , # 49                
#   "TCGA-LUAD" , # 27                
#   "TCGA-LUSC" , # 40                
#   # # "BRB"       , # 7                
#   # "TCGA-BRCA" , # 85                
#   # "TCGA-BLCA" , # 21                
#   # # "TCGA-CESC" , # 3                
#   # # "TCGA-CHOL" , # 9                
#   # "TCGA-COAD" , # 36                
#   # # "TCGA-DLBC" , # 0                
#   # # "TCGA-ESCA" , # 16                
#   # # "TCGA-LAML" , # 0                
#   # "TCGA-LIHC" , # 49                
#   # # "TCGA-MESO" , # 0                
#   # # "TCGA-PAAD" , # 10                
#   # # "TCGA-PCPG" , # 3                
#   # "TCGA-PRAD" , # 50                
#   # # "TCGA-READ" , # 7                
#   # # "TCGA-SARC" , # 2                
#   # # "TCGA-SKCM" , # 2                
#   # # "TCGA-STAD" , # 2                
#   # # "TCGA-TGCT" , # 0                
#   # "TCGA-THCA" , # 56                
#   # # "TCGA-THYM" , # 2                    
#   # # "TCGA-UCS"  , # 0                
#   # # "TCGA-UVM"  , # 0                
#   # # "TCGA-OV"   , # 0                     
#   NULL
# )


results = lapply(tcga_studies, function(tcga_id) {
  s = mreadRDS(paste0("~/projects/datashare/", tcga_id, "/study_preproc_", tcga_id, ".rds"))
  e = s$exp_grp
  idx = intersect(e[e$tissue_status == "normal",]$id_patient, e[e$tissue_status == "tumoral",]$id_patient)
  e = e[e$id_patient%in%idx,]
  tab1 = table(e[e$tissue_status == "normal",]$id_patient)
  tab2 = table(e[e$tissue_status == "tumoral",]$id_patient)
  idx = intersect(names(tab1)[tab1==1], names(tab2)[tab2==1])    
  e = e[e$id_patient%in%idx,]
  tab = table(e$id_patient, e$tissue_status)
  tab
  if (!all(tab==1)|nrow(tab)==0) {stop(paste0("Problem pairing samples for ", tcga_id))}

  qt = quantile(e$age)
  idx_yng = rownames(e)[e$age<=qt[2]]
  idx_old = rownames(e)[e$age>=qt[4]]
  individuals = list(all=rownames(e), young=idx_yng, old=idx_old)
  results = lapply(1:length(individuals), function(i, tcga_id) {
    print(paste0(tcga_id, " ", names(individuals[i])))
    idx = individuals[[i]]
    d = s$data[,idx]
    da = apply(d[,idx[e[idx,]$tissue_status=="tumoral"]], 1, mean) - apply(d[,idx[e[idx,]$tissue_status=="normal"]], 1, mean)
    da = sort(da)
    gsea_input = data.frame(probe=names(da), meth=da)
    rownames(gsea_input) = gsea_input$probe
    head(gsea_input)
    dim(gsea_input)
    # BOXPLOT
    results = lapply(names(grp_files), function(grp_file_name, tcga_id, gsea_input)  {
      grp_file = grp_files[[grp_file_name]]
      tmp_probes = intersect(read.table(grp_file)[,1], rownames(gsea_input))
      print(length(tmp_probes))
      # tmp_probes = sample(tmp_probes, 30)
      levs = c(grp_file_name, "BG")
      gsea_input$grp = levs[[2]]
      gsea_input[tmp_probes,]$grp = levs[[1]]
      gsea_input$grp = factor(gsea_input$grp, levels=levs)
      mw_pv = wilcox.test(gsea_input$meth[gsea_input$grp==levs[2]], gsea_input$meth[gsea_input$grp!=levs[2]])$p.value
      m = lm(gsea_input$meth~gsea_input$grp)
      beta = -m$coefficients[[2]]

      # if (i>0) { par(mar=c(2.1, 4.1, 4.1, .6)) } else { 	par(mar=c(5.1, 4.1, 4.1, .6)) }
      # main = ifelse(i>0, paste0("MW pv=", signif(mw_pv, 2)), paste0("Mann Whitney pv=", signif(mw_pv, 2)))
      # ylab = ifelse(i>0, "diff. meth.", "differential methylation")
      # las = ifelse(i>0, 1, 2)
      # tmpx = paste0(gsea_input$grp, " (", table(gsea_input$grp)[gsea_input$grp], " probes)")
      # tmpx = gsea_input$grp
      # beanplot::beanplot(gsea_input$meth ~ tmpx, what=c(1,1,0,0), main=main, ylab=ylab, las=las, yaxt="n", col=as.list(c(metacluster_cols[which(grp_files == grp_file)], "grey")))
      # # boxplot(gsea_input$meth ~ tmpx, what=c(1,1,0,0), main=main, ylab=ylab, las=las, yaxt="n", col=c(metacluster_cols[which(grp_files == grp_file)], "grey"), outline=FALSE)
      # axis(2, at=c(-.5,0,.5)) 

      return(data.frame(tcga_id=tcga_id, n=length(individuals[[i]])/2, grp=names(individuals[i]), beta=beta, pv=mw_pv, grp_file=grp_file))
    }, tcga_id, gsea_input)
    results = do.call(rbind, results)
    return(results)
  }, tcga_id)
  results = do.call(rbind, results)
  results$grp = factor(results$grp, names(individuals))
  return(results)
})
results = do.call(rbind, results)
results$grp_file = factor(results$grp_file, levels=grp_files)
levels(results$grp_file) = names(grp_files)
```

```{r child = '05_postprocgsea.Rmd', eval=TRUE}
```

```{r child = '05_postprocpancancer.Rmd', eval=TRUE}
```








# Process Information 

```{r, results="verbatim"}
info$stop_time = Sys.time()
info$exec_time = round(difftime(info$stop_time, info$start_time, units="mins"),1)
saveRDS(info, paste0("info_postproc.rds"))
print(info)
```
# Session Information

```{r, results="verbatim"}
sessionInfo()
```

