---
title: "Post processing `ewas`"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---


```{r echo=FALSE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
source("common.R")
```

## Figure

```{r figure ewas,fig.width=10, fig.height=10*2/3}
ewasfig = function()  {
  m0 = matrix(c(
      2,2,2,2,3,
      4,4,4,4,1,
      4,4,4,4,1,
      4,4,4,4,1,
      4,4,4,4,1,
  NULL), 5, byrow=TRUE)
  m0 = matrix(c(
      3,3,3,3,4,
      1,1,1,1,2,
      1,1,1,1,2,
      1,1,1,1,2,
      1,1,1,1,2,
  NULL), 5, byrow=TRUE)
  m0 = rbind(m0, m0 + max(m0))
  m = cbind(m0, m0+max(m0))
  m = cbind(m, m0+max(m))
  # m2 = m0
  # m2 = cbind(m2, m0+max(m2))
  # # m2 = cbind(m2, m0+max(m2))
  # # m2 = cbind(m2, m0+max(m2))
  # m = rbind(m, m2 + max(m))

  layout(m, respect=TRUE)

  # layout(matrix(c(
  # 1,1,2,2,
  # 1,1,2,2,
  # 3,4,5,6,
  # NULL), 3, byrow=TRUE), respect=TRUE)


  # par(mar=c(2, 2.5, 3.1, 2.5))
  # par(mar=c(5.1, 4.1, 4.1, 2.1))
  idx = rownames(ewas)#[ewas$pvbonf<0.01]
  x = ewas[idx, ]$beta
  y = ewas[idx, ]$lpv
  # d = density(y)
  xlab = "age effect (meth. prop. per year)"
  ylab = "-log10(pval)"
  main = paste0("methylation~age (", nrow(ewas), " probes)")
  # col = as.numeric(ewas$pvbonf<0.01)+ 1
  # plot(x, y, xlab=xlab, ylab=ylab, main=main, col=adjustcolor(col, alpha.f=.2), pch=16)
  # # points(x, y, col=adjustcolor(col, alpha.f=.5), pch=".")
  # legend("topleft", pch=16, col=1:2, c(paste0(">= 1% (", sum(!ewas$pvbonf<0.01), " probes)"), paste0("< 1% (", sum(ewas$pvbonf<0.01), " probes)")), title="Bonferroni")
  xlim = c(-0.005, 0.005)
  ylim = c(0, 80)
  tmp_i = 1
  custom_hm(x, y, xlim, ylim, xlab, ylab, main, col=1, LAYOUT=FALSE, LOGCOUNT=TRUE, wrapper_func=function(){abline(h=min(ewas[ewas$pvbonf<0.01, ]$lpv), col=2)}, letter=letters[tmp_i])
  tmp_i = tmp_i + 1

  idx = rownames(ewas)[ewas$pvbonf<0.01]
  xlab = "age effect (meth. prop. per year)"
  ylab = "mean methylation proportion"
  xlim = c(-0.003, 0.003)
  ylim = 0:1
  x = ewas[idx, ]$beta
  y = ewas[idx, ]$mean_beta
  col = "black"
  main = paste0("Bonferroni 1% (", length(idx), " probes)")
  custom_hm(x, y, xlim, ylim, xlab, ylab, main, col, LAYOUT=FALSE, letter=letters[tmp_i])
  tmp_i = tmp_i + 1

  # plot(x, y, xlab=xlab, ylab=ylab, main=main, col=adjustcolor(col, alpha.f=0.02*10000/length(idx)), pch=16, xlim=xlim, ylim=ylim)
  for (k in levels(ewas$Relation_to_Island)) {
    ylab = "mean meth. prop."
    # par(mar=c(2, 2.5, 3.1, 2.5))
    idx = rownames(ewas)[ewas$pvbonf<0.01]
    idx = intersect(idx, rownames(ewas)[ewas$Relation_to_Island%in% k])
    col = colpals_cpgdensity[k]
    x = ewas[idx, ]$beta
    y = ewas[idx, ]$mean_beta
    main=paste0(k, " (", length(idx), " probes)")
    # plot(x, y, xlab=xlab, ylab=ylab, main=main, col=adjustcolor(col, alpha.f=0.01*10000/length(idx)), pch=16, xlim=xlim, ylim=ylim)
    custom_hm(x, y, xlim, ylim, xlab, ylab, main, col, LAYOUT=FALSE, letter=letters[tmp_i])
    tmp_i = tmp_i + 1

  }
  # par(mar=c(5.1, 4.1, 4.1, 2.1))
  # stop("EFN")
}
ewasfig()
w = 10
pdf(paste0("fig_", gse, "_ewas.pdf"), width=w, height=w*2/3)
ewasfig()
dev.off()


# # pdf("ewasfig.pdf", width=8, height=6)
# png(paste0("fig_ewas_", gse, ".png"), width=1600, height=1200, res=150)
# ewasfig()
# dev.off()

```

```{r figure sup ewas,fig.width=10, fig.height=10*2/3}
ewasfigsup = function()  {
  # m0 = matrix(c(
  #     2,2,2,2,3,
  #     4,4,4,4,1,
  #     4,4,4,4,1,
  #     4,4,4,4,1,
  #     4,4,4,4,1,
  # NULL), 5, byrow=TRUE)
  # m0 = matrix(c(
  #     3,3,3,3,4,
  #     1,1,1,1,2,
  #     1,1,1,1,2,
  #     1,1,1,1,2,
  #     1,1,1,1,2,
  # NULL), 5, byrow=TRUE)
  # m0 = rbind(m0, m0 + max(m0))
  # m = cbind(m0, m0+max(m0))
  # m = cbind(m, m0+max(m))
  # # m2 = m0
  # # m2 = cbind(m2, m0+max(m2))
  # # # m2 = cbind(m2, m0+max(m2))
  # # # m2 = cbind(m2, m0+max(m2))
  # # m = rbind(m, m2 + max(m))

  # layout(m, respect=TRUE)

  idx = rownames(ewas)[ewas$pvbonf<0.01 & ewas$beta<0]
  idx = intersect(idx, rownames(ewas)[ewas$Relation_to_Island%in% "Island"])
  # idx = intersect(idx, rownames(ewas)[ewas$Relation_to_Island%in% c("Shore","Island")])
  xlab = "age effect (meth. prop. per year)"
  ylab = "mean methylation proportion"
  xlim = c(-0.003, 0.003)
  ylim = 0:1
  x = ewas[idx, ]$beta
  y = ewas[idx, ]$mean_beta
  col = "black"
  main = paste0("Bonferroni 1%, beta<0, Island, (", length(idx), " probes)")
  custom_hm(x, y, xlim, ylim, xlab, ylab, main, col)
}
ewasfigsup()
# w = 10
# pdf(paste0("figsup_", gse, "_ewas.pdf"), width=w, height=w*2/3)
# ewasfigsup()
# dev.off()


```

```{r eval=FALSE}
options(scipen = 0)
library(missMethyl)
# params
array.type = "EPIC"
plot.bias = FALSE
prior.prob = TRUE
anno = NULL
equiv.cpg = TRUE
fract.counts = TRUE
sig.genes = FALSE
# collection
# kegg = missMethyl:::.getKEGG()
go  = missMethyl:::.getGO()
head(go$idTable)
tmp_idTable = go$idTable
dim(tmp_idTable)
tmp_idTable = tmp_idTable[tmp_idTable$ONTOLOGY=="BP",]
dim(tmp_idTable)
collection = go$idList[tmp_idTable$GOID]
# reference and target
tmp_idx = rownames(ewas)[ewas$pvbonf<0.01 & ewas$beta<0]
tmp_idx = rownames(ewas)[ewas$pvbonf<0.01 & ewas$beta<0 & ewas$mean_beta>.5]
# tmp_idx = intersect(tmp_idx, rownames(ewas)[ewas$Relation_to_Island%in% "Island"])
tmp_idx = intersect(tmp_idx, rownames(ewas)[ewas$Relation_to_Island%in% c("Shore","Island")])
sig.cpg = tmp_idx
tmp_idx = rownames(ewas)
# tmp_idx = intersect(tmp_idx, rownames(ewas)[ewas$Relation_to_Island%in% "Island"])
tmp_idx = intersect(tmp_idx, rownames(ewas)[ewas$Relation_to_Island%in% c("Shore","Island")])
all.cpg = tmp_idx
genomic.features = "ALL" #, "TSS200", "TSS1500", "Body",         "1stExon", "3'UTR", "5'UTR", "ExonBnd"), sig.genes = FALSE)
genomic.features = "TSS200"
genomic.features = c("TSS200" , "TSS1500","5'UTR")
# Go!
if (!exists("mmissmethyl_gsameth")) {mmissmethyl_gsameth=memoise::memoise(missMethyl::gsameth)}
result <- mmissmethyl_gsameth(sig.cpg = sig.cpg, all.cpg = all.cpg, 
    collection = collection, array.type = array.type, 
    plot.bias = plot.bias, prior.prob = prior.prob, anno = anno, 
    equiv.cpg = equiv.cpg, fract.counts = fract.counts, 
    genomic.features = genomic.features, sig.genes = sig.genes)
result <- merge(tmp_idTable, result, by.x = "GOID", 
    by.y = "row.names")
rownames(result) <- result$PathwayID
result$TERM = substr(result$TERM, 1, 45)
head(result[order(result$P.DE),-(1:2)], 30)

missmethylfig = function (result) {
  layout(1, respect=TRUE)
  result = result[result$FDR<0.05,]
  tag_blacklist = c("cell", "of", "in", "to", "by")
  tab = sort(table(unlist(strsplit(result$TERM," "))))
  tab = tab[!names(tab) %in% tag_blacklist]
  wordcloud::wordcloud(names(tab), tab, main=k, min.freq=3)
}


missmethylfig(result)
pdf(paste0("fig_", gse, "_missmethyl.pdf"), width=6, height=6)
# png("missmethylfig.png", width=1600, height=1200, res=150)
missmethylfig(result)
dev.off()



```
