---
title: "Post processing `deepTools`"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---


### Heatmap (metaclusters)

```{r hm2, fig.width=9, fig.height=9}
# stop("EFN")
metaclusters = list(
  PRC2=c("cluster_3"), 
  ENHA=c("cluster_4"), 
  PROM=c("cluster_1", "cluster_2"), 
  # DEF1=c("cluster_6"),
  OTHR=c("cluster_5")
)

metaclusters = list(
  PRC2=paste0("cluster_", c(2)), 
  # ENHA=paste0("cluster_", c(4)), 
  PROM=paste0("cluster_", c(1)), 
  OTHR=paste0("cluster_", c(3))
)

metacluster_cols = c(
  "#19198c", 
  "#195eff", 
  "#36e6ba", 
  "#bde633", 
  "#ff7719", 
  NULL
)

metacluster_cols = c(
  "#19198c", 
  "#195eff", 
  # "#36e6ba", 
  "#6FCC6C", 
  "#ff9e19", 
  NULL
)

metacluster_cols = c(
  "#19198c", 
  "#19d8ff", 
  # "#36e6ba", 
  "#ffe819", 
  # "#ff9e19", 
  NULL
)


# Nom	Code Hex	Aperçu
# Original	#8AFF87	
# Plus foncé	#6FCC6C	
# Encore plus	#5AA859	
# Très foncé	#3F773E

names(metacluster_cols) = names(metaclusters)

bed_files = sapply(names(metaclusters), function(tag) {
  print(tag)
  locus_bed = clusters[clusters$V13%in% metaclusters[[tag]], 1:6]
  rownames(locus_bed) = locus_bed[,4]
  # tag = paste0("metacluster", paste0(substr(tag, 9, 9), collapse=""))
  locus_bed_file = write_bed_tss(locus_bed, tag=tag)
  # names(locus_bed_file) = tag
  locus_bed_file
})
  
# bed_files = c(
#   bed_files,
#   # write_bed_tss(bed_gain       , "gain"       ), 
#   # # write_bed_tss(bed_loss       , "loss"       ), 
#   # write_bed_tss(bed_losshypo     , "losshypo"       ), 
#   # write_bed_tss(bed_losshyper     , "losshyper"       ), 
#   # write_bed_tss(bed_rnd_island , "rnd_island" ), 
#   # write_bed_tss(bed_rnd_opensea, "rnd_opensea"), 
#   NULL
# )

samplesLabel = gsub("-human", "", paste(paste0(db$Experiment.target, "_", db$Biosample.term.name), collapse=" "))
samplesLabel = gsub("_lung", "", samplesLabel)
regionsLabel = gsub(".bed", "", paste(names(bed_files), collapse=" "))
prefix = paste0("metacluster_", gse)
mf = mget_mf_clust(bwfiles, bed_files, samplesLabel, prefix=prefix, othargs=paste0(" --refPointLabel 'loc' --legendLocation none --xAxisLabel ''  --colorMap RdBu --regionsLabel ", regionsLabel, " --heatmapWidth 5 --heatmapHeight 14"))
# plotHeatmap -m mt_metaclusters_GSE147740.txt.gz  --samplesLabel H3K4me3_lung H3K4me3_lung H3K4me1_lung H3K4me1_lung H3K4me1_lung H3K27me3_lung H3K27me3_lung -o hm_metaclusters_GSE147740.png --outFileSortedRegions hm_metaclusters_GSE147740.bed --refPointLabel "locus" --legendLocation none --xAxisLabel ""

```


![metaclusters](`r paste0("hm_", prefix, ".png")`)


### Inject cluster names in ewas df

```{r}
# cluster ENCODE
ewas$cluster = NA
for (bed_file_name in names(bed_files)) {
  locus_bed = read.table(bed_files[[bed_file_name]])
  idx = unique(unlist(fat_feats[locus_bed[,4],]$probes))
  ewas[idx,]$cluster = bed_file_name
}
```


### Clusters (metaclusters)


```{r, fig.width=9, fig.height=9}
plot_terrain(bed_files)
```

### Figure

```{r figure deepTools, fig.width=10*4/3, fig.height=10*2/3}
plot_predimethbiology = function() {

  # m0 = matrix(c(
  #     3,3,3,3,4,
  #     1,1,1,1,2,
  #     1,1,1,1,2,
  #     1,1,1,1,2,
  #     1,1,1,1,2,
  # NULL), 5, byrow=TRUE)
  # # m0 = rbind(m0, m0 + max(m0))
  # # m0 = cbind(m0, m0+max(m0))

  # m = matrix(c(
  #     1,1,1,1,1,1,1,1,1,1,
  #     1,1,1,1,1,1,1,1,1,1,
  #     1,1,1,1,1,1,1,1,1,1,
  #     1,1,1,1,1,1,1,1,1,1,
  #     1,1,1,1,1,1,1,1,1,1,
  #     1,1,1,1,1,1,1,1,1,1,
  #     1,1,1,1,1,1,1,1,1,1,
  #     1,1,1,1,1,1,1,1,1,1,
  #     1,1,1,1,1,1,1,1,1,1,
  #     1,1,1,1,1,1,1,1,1,1,
  # NULL), 10, byrow=TRUE)
  # m = cbind(m, m0+max(m))

  m = matrix(c(

      3,3,3,3,4, 5,5,5,5,5, 6,6,6,6,6,
      1,1,1,1,2, 5,5,5,5,5, 6,6,6,6,6,
      1,1,1,1,2, 5,5,5,5,5, 6,6,6,6,6,
      1,1,1,1,2, 5,5,5,5,5, 6,6,6,6,6,
      1,1,1,1,2, 5,5,5,5,5, 6,6,6,6,6,

      7,7,7,7,7,7,7, 12, 10,10,10,10,10,10,11,
      7,7,7,7,7,7,7, 12, 10,10,10,10,10,10,11,
      7,7,7,7,7,7,7, 12, 8,8,8,8,8,8,9,
      7,7,7,7,7,7,7, 12, 8,8,8,8,8,8,9,
      7,7,7,7,7,7,7, 12, 8,8,8,8,8,8,9,
      7,7,7,7,7,7,7, 12, 8,8,8,8,8,8,9,
      7,7,7,7,7,7,7, 12, 8,8,8,8,8,8,9,

  NULL), 12, byrow=TRUE)
  layout(m, respect=TRUE)


  # par(mar=c(2, 2.5, 3.1, 2.5))
  # par(mar=c(5.1, 4.1, 4.1, 2.1))
  idx = rownames(ewas)#[ewas$pvbonf<0.01]
  x = ewas[idx, ]$beta
  y = ewas[idx, ]$lpv
  # d = density(y)
  xlab = "age effect (meth. prop. per year)"
  ylab = "-log10(pval)"
  main = paste0("methylation~age (", nrow(ewas), " probes)")
  # col = as.numeric(ewas$pvbonf<0.01)+ 1
  # plot(x, y, xlab=xlab, ylab=ylab, main=main, col=adjustcolor(col, alpha.f=.2), pch=16)
  # # points(x, y, col=adjustcolor(col, alpha.f=.5), pch=".")
  # legend("topleft", pch=16, col=1:2, c(paste0(">= 1% (", sum(!ewas$pvbonf<0.01), " probes)"), paste0("< 1% (", sum(ewas$pvbonf<0.01), " probes)")), title="Bonferroni")
  xlim = c(-0.005, 0.005)
  ylim = c(0, 80)
  wrapper_func = function() {
    abline(h=min(ewas[ewas$pvbonf<0.01, ]$lpv), col=2)
    # widx = rownames(ewas)[ewas$predimeth==1]
    # wx = ewas[widx, ]$beta
    # wy = ewas[widx, ]$lpv
    # points(wx, wy, col=adjustcolor(2, alpha.f=0.3), pch=16)
  }


  l_it =  1
  custom_hm(x, y, xlim, ylim, xlab, ylab, main, col=1, LAYOUT=FALSE, LOGCOUNT=TRUE, wrapper_func=wrapper_func, letter=letters[l_it])
  l_it = l_it + 1






  options(scipen = 0)
  library(missMethyl)
  # params
  array.type = "EPIC"
  plot.bias = FALSE
  prior.prob = TRUE
  anno = NULL
  equiv.cpg = TRUE
  fract.counts = TRUE
  sig.genes = FALSE
  # collection
  # kegg = missMethyl:::.getKEGG()
  go  = missMethyl:::.getGO()
  head(go$idTable)
  tmp_idTable = go$idTable
  dim(tmp_idTable)
  tmp_idTable = tmp_idTable[tmp_idTable$ONTOLOGY=="BP",]
  dim(tmp_idTable)
  collection = go$idList[tmp_idTable$GOID]
  # reference and target
  sig.cpg = rownames(ewas)[ewas$pvbonf<0.01]
  # sig.cpg = rownames(ewas)[ewas$predimeth==1]
  all.cpg = rownames(ewas)
  genomic.features = "ALL" #, "TSS200", "TSS1500", "Body",         "1stExon", "3'UTR", "5'UTR", "ExonBnd"), sig.genes = FALSE)
  # Go!
  if (!exists("mmissmethyl_gsameth")) {mmissmethyl_gsameth=memoise::memoise(missMethyl::gsameth)}
  result <- mmissmethyl_gsameth(sig.cpg = sig.cpg, all.cpg = all.cpg, 
      collection = collection, array.type = array.type, 
      plot.bias = plot.bias, prior.prob = prior.prob, anno = anno, 
      equiv.cpg = equiv.cpg, fract.counts = fract.counts, 
      genomic.features = genomic.features, sig.genes = sig.genes)
  result <- merge(tmp_idTable, result, by.x = "GOID", 
      by.y = "row.names")
  rownames(result) <- result$PathwayID
  result$TERM_short = substr(result$TERM, 1, 35)
  result$EXP = result$N * length(sig.cpg) / length(all.cpg) 
  head(result[order(result$P.DE),-(1:3)], 30)

  # Sélection des 20 premiers termes les plus significatifs
  ntop=20
  top_res <- result[order(result$P.DE),][1:ntop,]
  top_res <- top_res[order(top_res$N),]

  # Préparation du graphique
  par(mar = c(5.1, 15, 4.1, 0.1))  # marges pour laisser de la place aux labels
  bp = barplot(
    top_res$N, 
    # -log10(top_res$P.DE), 
    names.arg = top_res$TERM,
    horiz = TRUE,
    las = 1,  # rotation des noms pour qu'ils soient horizontaux
    col = "lightgrey", 
    xlab = "#genes",
    main = paste0("Top ", ntop, " GO terms (BP)"),  
    border=0
  )
  barplot(top_res$DE,  horiz = TRUE,col = "darkgrey", add=TRUE, border=0, xaxt="n", yaxt="n")
  barplot(top_res$EXP, horiz = TRUE,col = "black",    add=TRUE, border=0, xaxt="n", yaxt="n")
  legend("bottomright", c("N", "DM", "EXP"), col=c("lightgrey", "darkgrey", "black"), pt.cex=2, pch=15)
  par(xpd = NA)
  text(top_res$N, bp, paste0("FDR=", signif(top_res$FDR, 1)), pos=4)
  par(xpd = FALSE)
  l = letters[l_it] ; l_it = l_it + 1
  put_a_letter(paste0(l))

  par(mar=c(5.1, 4.1, 4.1, 2.1))
  par(mar=c(0,0,0,0))
  result = result[result$FDR<0.05,]
  tag_blacklist = c("cell", "of", "in", "to", "by")
  tab = sort(table(unlist(strsplit(result$TERM," "))))
  tab = tab[!names(tab) %in% tag_blacklist]
  par(xpd = NA)
  if (length(tab)>0) {
    wordcloud::wordcloud(names(tab), tab, min.freq=3, use.r.layout=TRUE)
  } else {
    plot(0,0, main="no signif. term")
  }
  par(xpd = FALSE)
  par(mar=c(5.1, 4.1, 4.1, 2.1))
  l = letters[l_it] ; l_it = l_it + 1
  put_a_letter(paste0(l))







  par(mar=c(0, 0, 0, 0))
  img = png::readPNG(paste0("hm_", prefix, ".png"))
  plot(NA, NA, xlim=0:1, ylim=0:1, axes=FALSE, main=paste0(""), xlab="", ylab="")
  addImg(img, x=.5,y=.5,width=1)
  par(mar=c(5.1, 4.1, 4.1, 2.1))  
  l = letters[l_it] ; l_it = l_it + 1
  put_a_letter(paste0(l))
  par(mar=c(5.1, 4.1, 4.1, 2.1))




  tmp_df = ewas[ewas$predimeth==1,]
  wx = tmp_df$beta
  wy = tmp_df$lpv
  wy[wy>150]=150
  xlab = "age effect (meth. prop. per year)"
  ylab = "-log10(pval)"
  main = paste0("methylation~age (", nrow(tmp_df), " PrediMeth probes)")


  levs = names(bed_files)
  tmp_df$cluster = factor(tmp_df$cluster, levels=levs)

  par(mar=c(5.1, 4.1, 0, 0))
  plot(wx, wy, col=adjustcolor(metacluster_cols[tmp_df$cluster], alpha.f=0.66), pch=16, xlab=xlab, ylab=ylab)
  legend("topleft", paste0(levels(tmp_df$cluster), " (", sapply(levels(tmp_df$cluster), function(k){sum(tmp_df$cluster==k)}), " probes)"), col=metacluster_cols[levels(tmp_df$cluster)], pch=16)

  
  par(mar=c(5.1, 0, 0, .1))
  par(xpd = NA)
  den = density(wy)
  denx = den$x
  deny = den$y



  plot(deny, denx, type="l", frame.plot=FALSE, yaxs="i", xaxt="n", yaxt="n", xlab="", ylab="", col="grey", lwd=1, lty=3)
  for (k in levs) {
    den2 = density(wy[tmp_df$cluster==k], bw = den$bw)
    denx = den2$x
    deny = den2$y/nrow(tmp_df)*sum(tmp_df$cluster==k)
    lines(deny, denx, type="l", col=metacluster_cols[[k]], lwd=3) 
  }
  par(xpd = FALSE)

  par(mar=c(0, 4.1, 4.1, 0))
  par(xpd = NA)
  den = density(wx)
  denx = den$x
  deny = den$y
  plot(denx, deny, type="l", frame.plot=FALSE, xaxs="i", xaxt="n", yaxt="n", xlab="", ylab="", main=main, col="grey", lwd=1, lty=3) 
  for (k in levs) {
    den2 = density(wx[tmp_df$cluster==k], bw = den$bw)
    denx = den2$x
    deny = den2$y/nrow(tmp_df)*sum(tmp_df$cluster==k)
    lines(denx, deny, type="l", col=metacluster_cols[[k]], lwd=3) 
  }
  l = letters[l_it] ; l_it = l_it + 1
  put_a_letter(paste0(l))
  par(xpd = FALSE)


  par(mar=c(5.1, 4.1, 4.1, 2.1))
}
# stop("EFN")
plot_predimethbiology()
# png(paste0("fig_predimethbiology_", gse, ".png"), width=1800, height=900, res=150) 
w = 40/3
pdf(paste0("fig_", gse, "_predimethbiology.pdf"), width=w, height=w*12/15) 
plot_predimethbiology()
dev.off()
```












```{r results="verbatim"}
probes_per_bed_files = lapply(bed_files, function(bed_file) {  
  print(bed_file)
  locus_bed = read.table(bed_file)
  rownames(locus_bed) = locus_bed[,4]
  head(locus_bed)
  tmp_probes = unique(unlist(fat_feats[rownames(locus_bed),]$probes))
  return(tmp_probes)
})
names(probes_per_bed_files) = bed_files
sapply(probes_per_bed_files, length)

probes_per_bed_files$bonferroni01 = probes_bonf01

ewas$metacluster = NA
for (k in names(bed_files)) {
  foo = probes_per_bed_files[[bed_files[[k]]]]
  ewas[foo, ]$metacluster = k
}

pv_tab = table(ewas$Relation_to_Island, ewas$metacluster)
for (k1 in levels(ewas$Relation_to_Island)) {
  for (k in names(bed_files)) {
    print(k)
    tab = table(ewas$Relation_to_Island%in%k1, ewas$metacluster%in%k, useNA="ifany")
    print(tab)
    ptab = prop.table(tab, margin=2)
    print(ptab)
    chi2 = chisq.test(tab)
    pv_tab[k1, k] = signif(chi2$p.value,3)
    print(chi2)
  }
}
prop.table(table(ewas$Relation_to_Island, ewas$metacluster), margin=1)
prop.table(table(ewas$Relation_to_Island, ewas$metacluster, useNA="ifany"), margin=2)
pv_tab

```

```{r eval=FALSE}

# k1=c("Island", "Shore", "Shelf")
# k1 = "OpenSea"
k1="Island"
k2="OTHR"
# k2="ENHA"
main=paste0(k1, "/", k2)
idx = rownames(ewas)[ewas$Relation_to_Island%in%k1 & ewas$metacluster%in%k2]
sig.cpg = idx
all.cpg = rownames(ewas)

genomic.features = "ALL" #, "TSS200", "TSS1500", "Body",         "1stExon", "3'UTR", "5'UTR", "ExonBnd"), sig.genes = FALSE)
# Go!
if (!exists("mmissmethyl_gsameth")) {mmissmethyl_gsameth=memoise::memoise(missMethyl::gsameth)}
result <- mmissmethyl_gsameth(sig.cpg = sig.cpg, all.cpg = all.cpg, 
    collection = collection, array.type = array.type, 
    plot.bias = plot.bias, prior.prob = prior.prob, anno = anno, 
    equiv.cpg = equiv.cpg, fract.counts = fract.counts, 
    genomic.features = genomic.features, sig.genes = sig.genes)
result <- merge(tmp_idTable, result, by.x = "GOID", 
    by.y = "row.names")
rownames(result) <- result$PathwayID
result$TERM_short = substr(result$TERM, 1, 35)
result$EXP = result$N * length(sig.cpg) / length(all.cpg) 
head(result[order(result$P.DE),-(1:3)], 30)

col = 1
xlab = "age effect (meth. prop. per year)"
ylab = "mean meth. prop."
xlim = c(-0.003, 0.003)
ylim = 0:1
x = ewas[idx, ]$beta
y = ewas[idx, ]$mean_beta
main=paste0(main, " (", length(idx), " probes)")
custom_hm(x, y, xlim, ylim, xlab, ylab, main, col)
missmethylfig(result)


```

```{r}



predimeth_probes = rownames(ewas)[ewas$predimeth==1]
predimeth_probes_per_bed_files = lapply(names(probes_per_bed_files), function(bed_file) {  
  tmp_pbs = predimeth_probes[predimeth_probes %in% probes_per_bed_files[[bed_file]]]
  write.table(tmp_pbs, paste0("predimeth_", bed_file, ".grp"), quote=FALSE, row.names=FALSE, col.names=FALSE)
  return(tmp_pbs)
})
names(predimeth_probes_per_bed_files) = names(probes_per_bed_files)
sapply(predimeth_probes_per_bed_files, length)

grp_files = sapply(names(bed_files), function(k) {
  tmp_pbs = predimeth_probes_per_bed_files[[paste0(k, ".bed")]]
  grp_file = paste0("predimeth_", strsplit(k, "\\.")[[1]][1], ".grp")
  print(grp_file)
  write.table(tmp_pbs, grp_file, quote=FALSE, row.names=FALSE, col.names=FALSE)
  return(grp_file)
})

# grp_file = "rnd5000_island.grp"
# write.table(idx_rnd_island, grp_file, quote=FALSE, row.names=FALSE, col.names=FALSE)
# grp_files["rndisland"] = grp_file

# grp_file = "rnd5000_opensea.grp"
# write.table(idx_rnd_opensea, grp_file, quote=FALSE, row.names=FALSE, col.names=FALSE)
# grp_files["rndopensea"] = grp_file


chi = data.frame(
  all = sapply(probes_per_bed_files, length),
  pdm = sapply(predimeth_probes_per_bed_files, length),
  exp = round(sapply(probes_per_bed_files, length) / length(probes_per_bed_files$bon) * length(predimeth_probes_per_bed_files$bon))
)
chi$rat = signif(chi$pdm / chi$all,3)
chi$enr = signif(chi$pdm / chi$exp,3)
chi
```


# Associated genes


```{r eval=TRUE}
genome = "hg38"
if (!exists("mannotatr_build_annotations")) {mannotatr_build_annotations = memoise::memoise(annotatr::build_annotations)}
annotations = mannotatr_build_annotations(genome=genome, annotations=paste0(genome, c("_basicgenes")))

genes = lapply(bed_files, function(bed_file) {  
  print(bed_file)
  tmp_regions = annotatr::read_regions(con=bed_file, genome=genome, format="bed")
  # Intersect the regions we read in with the annotations
  tmp_regions_annotated = annotatr::annotate_regions(
      regions = tmp_regions,
      annotations = annotations,
      ignore.strand = TRUE,
      quiet = FALSE)
  # A GRanges object is returned
  tmp_regions_annotated = data.frame(tmp_regions_annotated)
  dim(tmp_regions_annotated)
  # dedup
  tmp_regions_annotated$annot.type = factor(tmp_regions_annotated$annot.type, levels=paste0(genome, c("_genes_promoters", "_genes_1to5kb", "_genes_5UTRs", "_genes_exons", "_genes_introns", "_genes_3UTRs")))
  table(tmp_regions_annotated$annot.type)
  table(tmp_regions_annotated$name)
  tag = "hg38_genes_promoters"
  genes = unique(tmp_regions_annotated[tmp_regions_annotated$annot.type%in%tag & !is.na(tmp_regions_annotated$annot.symbol),]$annot.symbol)
  gene_filename = paste0(tag, "_", bed_file, ".txt")
  write.table(genes, gene_filename, quote=FALSE, row.names=FALSE, col.names=FALSE)
  print(paste0("genes writen in ", gene_filename))
  return(genes)
})
names(genes) = bed_files
```





